{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"name":"python","version":"3.10.13","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"},"kaggle":{"accelerator":"gpu","dataSources":[{"sourceId":8797460,"sourceType":"datasetVersion","datasetId":5289950}],"dockerImageVersionId":30733,"isInternetEnabled":true,"language":"python","sourceType":"notebook","isGpuEnabled":true}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load\n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n\n# Input data files are available in the read-only \"../input/\" directory\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))\n\n# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session","metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","execution":{"iopub.status.busy":"2024-06-27T17:58:59.027834Z","iopub.execute_input":"2024-06-27T17:58:59.028228Z","iopub.status.idle":"2024-06-27T17:58:59.405107Z","shell.execute_reply.started":"2024-06-27T17:58:59.028190Z","shell.execute_reply":"2024-06-27T17:58:59.404238Z"},"trusted":true},"execution_count":1,"outputs":[{"name":"stdout","text":"/kaggle/input/bsebse/BSE102_Lecture 3_Chemical basis of life.pdf\n","output_type":"stream"}]},{"cell_type":"code","source":"pip install pymupdf","metadata":{"execution":{"iopub.status.busy":"2024-06-27T17:58:59.406503Z","iopub.execute_input":"2024-06-27T17:58:59.406840Z","iopub.status.idle":"2024-06-27T17:59:11.439116Z","shell.execute_reply.started":"2024-06-27T17:58:59.406815Z","shell.execute_reply":"2024-06-27T17:59:11.437838Z"},"trusted":true},"execution_count":2,"outputs":[{"name":"stdout","text":"Requirement already satisfied: pymupdf in /opt/conda/lib/python3.10/site-packages (1.24.7)\nRequirement already satisfied: PyMuPDFb==1.24.6 in /opt/conda/lib/python3.10/site-packages (from pymupdf) (1.24.6)\nNote: you may need to restart the kernel to use updated packages.\n","output_type":"stream"}]},{"cell_type":"code","source":"pip install einops flash_attn","metadata":{"execution":{"iopub.status.busy":"2024-06-27T17:59:11.440457Z","iopub.execute_input":"2024-06-27T17:59:11.440783Z","iopub.status.idle":"2024-06-27T17:59:23.461885Z","shell.execute_reply.started":"2024-06-27T17:59:11.440755Z","shell.execute_reply":"2024-06-27T17:59:23.460688Z"},"trusted":true},"execution_count":3,"outputs":[{"name":"stdout","text":"Requirement already satisfied: einops in /opt/conda/lib/python3.10/site-packages (0.8.0)\nRequirement already satisfied: flash_attn in /opt/conda/lib/python3.10/site-packages (2.5.9.post1)\nRequirement already satisfied: torch in /opt/conda/lib/python3.10/site-packages (from flash_attn) (2.1.2)\nRequirement already satisfied: filelock in /opt/conda/lib/python3.10/site-packages (from torch->flash_attn) (3.13.1)\nRequirement already satisfied: typing-extensions in /opt/conda/lib/python3.10/site-packages (from torch->flash_attn) (4.9.0)\nRequirement already satisfied: sympy in /opt/conda/lib/python3.10/site-packages (from torch->flash_attn) (1.12.1)\nRequirement already satisfied: networkx in /opt/conda/lib/python3.10/site-packages (from torch->flash_attn) (3.2.1)\nRequirement already satisfied: jinja2 in /opt/conda/lib/python3.10/site-packages (from torch->flash_attn) (3.1.2)\nRequirement already satisfied: fsspec in /opt/conda/lib/python3.10/site-packages (from torch->flash_attn) (2024.3.1)\nRequirement already satisfied: MarkupSafe>=2.0 in /opt/conda/lib/python3.10/site-packages (from jinja2->torch->flash_attn) (2.1.3)\nRequirement already satisfied: mpmath<1.4.0,>=1.1.0 in /opt/conda/lib/python3.10/site-packages (from sympy->torch->flash_attn) (1.3.0)\nNote: you may need to restart the kernel to use updated packages.\n","output_type":"stream"}]},{"cell_type":"code","source":"pip install python-docx","metadata":{"execution":{"iopub.status.busy":"2024-06-27T17:59:23.464910Z","iopub.execute_input":"2024-06-27T17:59:23.465261Z","iopub.status.idle":"2024-06-27T17:59:35.482419Z","shell.execute_reply.started":"2024-06-27T17:59:23.465232Z","shell.execute_reply":"2024-06-27T17:59:35.481389Z"},"trusted":true},"execution_count":4,"outputs":[{"name":"stdout","text":"Requirement already satisfied: python-docx in /opt/conda/lib/python3.10/site-packages (1.1.2)\nRequirement already satisfied: lxml>=3.1.0 in /opt/conda/lib/python3.10/site-packages (from python-docx) (5.2.2)\nRequirement already satisfied: typing-extensions>=4.9.0 in /opt/conda/lib/python3.10/site-packages (from python-docx) (4.9.0)\nNote: you may need to restart the kernel to use updated packages.\n","output_type":"stream"}]},{"cell_type":"code","source":"import os\nimport fitz  \nfrom PIL import Image, ImageDraw, ImageFont\nimport io\nfrom io import BytesIO\nimport pytesseract  \nimport torch \nimport requests  \nfrom transformers import Blip2Processor, Blip2ForConditionalGeneration, PegasusForConditionalGeneration, PegasusTokenizer, AutoProcessor, AutoModelForCausalLM \nfrom docx import Document \nfrom docx.shared import Inches  \nimport tempfile  \nfrom transformers import T5Tokenizer, T5ForConditionalGeneration\nfrom transformers import BartForConditionalGeneration, BartTokenizer\nimport nltk\nfrom nltk.corpus import stopwords\nfrom nltk.tokenize import word_tokenize\nfrom transformers import BartTokenizer\nimport numpy as np \nimport pandas as pd \nnltk.download('punkt')\nnltk.download('stopwords')\n","metadata":{"execution":{"iopub.status.busy":"2024-06-27T17:59:35.491300Z","iopub.execute_input":"2024-06-27T17:59:35.491602Z","iopub.status.idle":"2024-06-27T17:59:42.422621Z","shell.execute_reply.started":"2024-06-27T17:59:35.491576Z","shell.execute_reply":"2024-06-27T17:59:42.421682Z"},"trusted":true},"execution_count":6,"outputs":[{"name":"stderr","text":"2024-06-27 17:59:38.936712: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n2024-06-27 17:59:38.936762: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n2024-06-27 17:59:38.938170: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n","output_type":"stream"},{"name":"stdout","text":"[nltk_data] Downloading package punkt to /usr/share/nltk_data...\n[nltk_data]   Package punkt is already up-to-date!\n[nltk_data] Downloading package stopwords to /usr/share/nltk_data...\n[nltk_data]   Package stopwords is already up-to-date!\n","output_type":"stream"},{"execution_count":6,"output_type":"execute_result","data":{"text/plain":"True"},"metadata":{}}]},{"cell_type":"code","source":"device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")","metadata":{"execution":{"iopub.status.busy":"2024-06-27T17:59:42.423896Z","iopub.execute_input":"2024-06-27T17:59:42.424439Z","iopub.status.idle":"2024-06-27T17:59:42.429170Z","shell.execute_reply.started":"2024-06-27T17:59:42.424409Z","shell.execute_reply":"2024-06-27T17:59:42.428254Z"},"trusted":true},"execution_count":7,"outputs":[]},{"cell_type":"code","source":"model = AutoModelForCausalLM.from_pretrained(\"microsoft/Florence-2-large\", trust_remote_code=True).to(device)\nprocessor = AutoProcessor.from_pretrained(\"microsoft/Florence-2-large\", trust_remote_code=True)","metadata":{"execution":{"iopub.status.busy":"2024-06-27T17:59:42.430324Z","iopub.execute_input":"2024-06-27T17:59:42.431097Z","iopub.status.idle":"2024-06-27T17:59:51.293901Z","shell.execute_reply.started":"2024-06-27T17:59:42.431061Z","shell.execute_reply":"2024-06-27T17:59:51.292865Z"},"trusted":true},"execution_count":8,"outputs":[{"name":"stderr","text":"/opt/conda/lib/python3.10/site-packages/torch/_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()\n  return self.fget.__get__(instance, owner)()\n","output_type":"stream"}]},{"cell_type":"code","source":"def run_example(image, task_prompt, text_input=None):\n    if text_input is None:\n        prompt = task_prompt\n    else:\n        prompt = task_prompt + text_input\n    inputs = processor(text=prompt, images=image, return_tensors=\"pt\").to(device)\n    generated_ids = model.generate(\n        input_ids=inputs[\"input_ids\"],\n        pixel_values=inputs[\"pixel_values\"],\n        max_new_tokens=1024,\n        num_beams=3\n    )\n    generated_text = processor.batch_decode(generated_ids, skip_special_tokens=False)[0]\n    parsed_answer = processor.post_process_generation(generated_text, task=task_prompt, image_size=(image.width, image.height))\n    caption = parsed_answer[\"<MORE_DETAILED_CAPTION>\"]\n    return caption","metadata":{"execution":{"iopub.status.busy":"2024-06-27T17:59:51.295226Z","iopub.execute_input":"2024-06-27T17:59:51.295528Z","iopub.status.idle":"2024-06-27T17:59:51.302427Z","shell.execute_reply.started":"2024-06-27T17:59:51.295501Z","shell.execute_reply":"2024-06-27T17:59:51.301563Z"},"trusted":true},"execution_count":9,"outputs":[]},{"cell_type":"code","source":"def pdf_doc(pdf_document):\n    doc = fitz.open(pdf_document)\n\n    page_text_list = []\n    for page_number in range(len(doc)):\n        page = doc.load_page(page_number)  \n        image_list = page.get_images(full=True)  \n\n        page_text = page.get_text()\n        page_text_combined = page_text\n\n        \n        for img_index, img in enumerate(image_list):\n            xref = img[0] \n            base_image = doc.extract_image(xref)  \n            image_bytes = base_image[\"image\"]\n            \n            image = Image.open(io.BytesIO(image_bytes)).convert('RGB')\n            \n            prompt = \"<MORE_DETAILED_CAPTION>\"\n            generated_caption = run_example(image, prompt)\n\n            combined_text = f\"\\n\\nImage {img_index + 1}:\\nCaption: {generated_caption}\\n\"\n\n            page_text_combined += combined_text\n\n        page_text_list.append(page_text_combined)\n    return page_text_list","metadata":{"execution":{"iopub.status.busy":"2024-06-27T17:59:51.305697Z","iopub.execute_input":"2024-06-27T17:59:51.305989Z","iopub.status.idle":"2024-06-27T17:59:51.314362Z","shell.execute_reply.started":"2024-06-27T17:59:51.305964Z","shell.execute_reply":"2024-06-27T17:59:51.313508Z"},"trusted":true},"execution_count":10,"outputs":[]},{"cell_type":"code","source":"def docx_doc(docx_document):\n    doc = Document(docx_document)\n    doc_text_list = []\n    for para in doc.paragraphs:\n        doc_text_list.append(para.text)\n\n        for run in para.runs:\n            if run.element.xml.find(\"pic:blipFill\") != -1:\n                image = run.element.xpath(\".//a:blip\")[0]\n                image_data = image.get(\"{http://schemas.openxmlformats.org/officeDocument/2006/relationships}embed\")\n                image_part = doc.part.related_parts[image_data]\n                image_bytes = image_part.blob\n\n                image = Image.open(io.BytesIO(image_bytes)).convert('RGB')\n                prompt = \"<MORE_DETAILED_CAPTION>\"\n                generated_caption = run_example(image, prompt)\n                combined_text = f\"Caption: {generated_caption}\\n\"\n                doc_text_list.append(combined_text)\n    return doc_text_list\n\n    \n\n","metadata":{"execution":{"iopub.status.busy":"2024-06-27T17:59:51.315695Z","iopub.execute_input":"2024-06-27T17:59:51.315991Z","iopub.status.idle":"2024-06-27T17:59:51.327277Z","shell.execute_reply.started":"2024-06-27T17:59:51.315959Z","shell.execute_reply":"2024-06-27T17:59:51.326512Z"},"trusted":true},"execution_count":11,"outputs":[]},{"cell_type":"code","source":"def determine_file_type(file_path):\n    _, file_extension = os.path.splitext(file_path)\n    if file_extension.lower() == '.docx':\n        return 'docx'\n    elif file_extension.lower() == '.pdf':\n        return 'pdf'\n    else:\n        return 'unknown'","metadata":{"execution":{"iopub.status.busy":"2024-06-27T17:59:51.328369Z","iopub.execute_input":"2024-06-27T17:59:51.328634Z","iopub.status.idle":"2024-06-27T17:59:51.339335Z","shell.execute_reply.started":"2024-06-27T17:59:51.328611Z","shell.execute_reply":"2024-06-27T17:59:51.338383Z"},"trusted":true},"execution_count":12,"outputs":[]},{"cell_type":"code","source":"def get_text(path):\n    text_each_page=[]\n    file_type=determine_file_type(path)\n    if file_type=='pdf':\n        text_each_page=pdf_doc(path)\n    elif file_type=='docx':\n        text_each_page=docx_doc(path)\n    else :\n        print(\"invalid file type\")\n    return text_each_page","metadata":{"execution":{"iopub.status.busy":"2024-06-27T17:59:51.340465Z","iopub.execute_input":"2024-06-27T17:59:51.341116Z","iopub.status.idle":"2024-06-27T17:59:51.348817Z","shell.execute_reply.started":"2024-06-27T17:59:51.341083Z","shell.execute_reply":"2024-06-27T17:59:51.347902Z"},"trusted":true},"execution_count":13,"outputs":[]},{"cell_type":"code","source":"def preprocess_text(text):\n    text = text.lower()\n    tokens = word_tokenize(text)\n    stop_words = set(stopwords.words('english'))\n    tokens = [word for word in tokens if word.isalpha() and word not in stop_words]\n    return ' '.join(tokens)\n","metadata":{"execution":{"iopub.status.busy":"2024-06-27T17:59:51.349834Z","iopub.execute_input":"2024-06-27T17:59:51.350137Z","iopub.status.idle":"2024-06-27T17:59:51.357327Z","shell.execute_reply.started":"2024-06-27T17:59:51.350100Z","shell.execute_reply":"2024-06-27T17:59:51.356470Z"},"trusted":true},"execution_count":14,"outputs":[]},{"cell_type":"code","source":"def process_text_in_sentence_chunks(path, max_chunk_size):\n    text=('\\n'.join(page for page in get_text(path)))\n    sentences = text.replace(\"\\n\", \" \") \n    sentences = nltk.sent_tokenize(sentences)\n\n    current_chunk = \"\"\n    text_list = []\n    for sentence in sentences:\n        if len(current_chunk) + len(sentence) <= max_chunk_size:\n            current_chunk += \" \" + sentence\n        else:\n              text_list.append(current_chunk.strip())\n              current_chunk = sentence\n\n    if current_chunk:\n        text_list.append(current_chunk.strip())\n    return text_list","metadata":{"execution":{"iopub.status.busy":"2024-06-27T17:59:51.358350Z","iopub.execute_input":"2024-06-27T17:59:51.358622Z","iopub.status.idle":"2024-06-27T17:59:51.369996Z","shell.execute_reply.started":"2024-06-27T17:59:51.358588Z","shell.execute_reply":"2024-06-27T17:59:51.369217Z"},"trusted":true},"execution_count":15,"outputs":[]},{"cell_type":"markdown","source":"# Models****","metadata":{}},{"cell_type":"code","source":"model_name = 't5-large'\ntokenizer1 = T5Tokenizer.from_pretrained(model_name)\nmodel1 = T5ForConditionalGeneration.from_pretrained(model_name)\n\ndevice = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\nmodel1.to(device)\n\ndef generate_summary_t5(text, max_length=60):\n    preprocess_text = text.strip().replace(\"\\n\", \" \")\n    t5_prepared_text = \"summarize: \" + preprocess_text\n\n    tokenized_text = tokenizer1.encode(t5_prepared_text, return_tensors=\"pt\", max_length=512, truncation=True)\n    tokenized_text = tokenized_text.to(device)\n    summary_ids = model1.generate(tokenized_text, max_length=max_length, min_length=5, length_penalty=2.0, num_beams=4, early_stopping=True)\n    summary = tokenizer1.decode(summary_ids[0], skip_special_tokens=True)\n\n    return summary","metadata":{"execution":{"iopub.status.busy":"2024-06-27T17:59:51.371147Z","iopub.execute_input":"2024-06-27T17:59:51.371637Z","iopub.status.idle":"2024-06-27T17:59:54.724088Z","shell.execute_reply.started":"2024-06-27T17:59:51.371611Z","shell.execute_reply":"2024-06-27T17:59:54.723030Z"},"trusted":true},"execution_count":16,"outputs":[{"name":"stderr","text":"You are using the default legacy behaviour of the <class 'transformers.models.t5.tokenization_t5.T5Tokenizer'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565\nSpecial tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.\n","output_type":"stream"}]},{"cell_type":"code","source":"model_name = \"facebook/bart-large-cnn\"\ntokenizer2 = BartTokenizer.from_pretrained(model_name)\nmodel2 = BartForConditionalGeneration.from_pretrained(model_name)\n\ndevice = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\nmodel2.to(device)\n\n\ndef generate_summary_bart(input_text):\n    inputs = tokenizer2(input_text, return_tensors=\"pt\", max_length=1024, truncation=True).to(\"cuda\")\n    summary_ids = model2.generate(inputs['input_ids'], \n                                 max_length=80, \n                                 min_length=5, \n                                 length_penalty=2.0, \n                                 num_beams=4, \n                                 early_stopping=True)\n    summary = tokenizer2.decode(summary_ids[0], skip_special_tokens=True)\n    return summary\n","metadata":{"execution":{"iopub.status.busy":"2024-06-27T17:59:54.725267Z","iopub.execute_input":"2024-06-27T17:59:54.725568Z","iopub.status.idle":"2024-06-27T17:59:57.482420Z","shell.execute_reply.started":"2024-06-27T17:59:54.725542Z","shell.execute_reply":"2024-06-27T17:59:57.481462Z"},"trusted":true},"execution_count":17,"outputs":[]},{"cell_type":"code","source":"model_name = \"google/pegasus-large\"\ntokenizer3 = PegasusTokenizer.from_pretrained(model_name)\nmodel3 = PegasusForConditionalGeneration.from_pretrained(model_name)\n\n\ndef generate_summary_pegasus(input_text, max_input_length=1024, max_summary_length=60, min_summary_length=15):\n    inputs = tokenizer3(input_text, return_tensors=\"pt\", max_length=max_input_length, truncation=True)\n    summary_ids = model3.generate(inputs['input_ids'], \n                                 max_length=max_summary_length,\n                                 min_length=min_summary_length,  \n                                 num_beams=4, \n                                 length_penalty=2.0, \n                                 early_stopping=True)\n    summary = tokenizer3.decode(summary_ids[0], skip_special_tokens=True)\n    return summary","metadata":{"execution":{"iopub.status.busy":"2024-06-27T17:59:57.483800Z","iopub.execute_input":"2024-06-27T17:59:57.484881Z","iopub.status.idle":"2024-06-27T18:00:07.171656Z","shell.execute_reply.started":"2024-06-27T17:59:57.484804Z","shell.execute_reply":"2024-06-27T18:00:07.170897Z"},"trusted":true},"execution_count":18,"outputs":[{"name":"stderr","text":"Some weights of PegasusForConditionalGeneration were not initialized from the model checkpoint at google/pegasus-large and are newly initialized: ['model.decoder.embed_positions.weight', 'model.encoder.embed_positions.weight']\nYou should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.\n","output_type":"stream"}]},{"cell_type":"code","source":"path=\"/kaggle/input/bsebse/BSE102_Lecture 3_Chemical basis of life.pdf\"\n","metadata":{"execution":{"iopub.status.busy":"2024-06-27T18:00:07.172809Z","iopub.execute_input":"2024-06-27T18:00:07.173181Z","iopub.status.idle":"2024-06-27T18:00:07.177366Z","shell.execute_reply.started":"2024-06-27T18:00:07.173144Z","shell.execute_reply":"2024-06-27T18:00:07.176462Z"},"trusted":true},"execution_count":19,"outputs":[]},{"cell_type":"code","source":"input_text=process_text_in_sentence_chunks(path,512)\nfor text in input_text:\n    print(generate_summary_t5(text))","metadata":{"execution":{"iopub.status.busy":"2024-06-27T18:00:07.178643Z","iopub.execute_input":"2024-06-27T18:00:07.179084Z","iopub.status.idle":"2024-06-27T18:02:31.462764Z","shell.execute_reply.started":"2024-06-27T18:00:07.179049Z","shell.execute_reply":"2024-06-27T18:02:31.461747Z"},"trusted":true},"execution_count":20,"outputs":[{"name":"stdout","text":"BSE 102 Bioscience Spring 2024 The Chemical Basis of Life Instructor: Dr. sunil Kumar boda Assistant professor department of biosciences & biomedical engineering IIT Indore.\nthe book is surrounded by a banner with the words \"IIT Indore\" written in Hindi. below the book, there are two lines of text in Hindi, which translates to \"Indian Institute of technology Indore\" the emblem is a symbol of knowledge and knowledge\nelements are substances that cannot be further broken down to other substances by ordinary chemical means. compounds are substances containing two or more elements combined in a fixed ratio.\n2Na + Cl2 2 NaCl; (both Na and Cl are corrosive, but. NaCl is edible!)\nmajor elements in the human body: C, O, H, N, Ca, P (make up 99% of body weight) minor elements ( 1 wt%): K, S, Na, Cl, Mg (for nerve signaling and chemical reactions) trace elements (\nthe infographic shows a man standing with his back to the viewer. he is wearing a red t-shirt, black shorts, and colorful socks. the infographic is divided into three sections.\nthe chart also includes information about the different elements and their properties. trace elements are common additives to food and water. iodine is an essential component of the hormones thyroxine (T4 - inactive) and triiodoth\niodine deficiency leads to abnormal swelling of the thyroid gland leading to a condition called goiter. most serious effects of iodine deficiency occur during fetal development and childhood.\niron (Fe) makes up only 0.004 wt% of body weight. it is vital for energy processing and transport of oxygen. iron rich foods include meat, fish, poultry, eggs, green, legumes, raisins.\nexcess fluoride in drinking water can cause fluorosis, a condition of brown stained teeth. excess/ overload of trace elements can lead to accidental fatal poisoning. recommended dietary allowances (RDA) of trace elements are based on gender (male/ female\nthe woman appears to be in her late 60s or early 70s. the text on the image reads \"goiter, a symptom of iodine deficiency\" the image shows a variety of food items that are rich in iron.\nimage shows a piece of raw meat, possibly pork or beef. the meat is pink in color and looks to be seasoned with herbs and spices. the background is white.\nmouthwash and toothpaste are examples of dietary deficiencies. a table lists the different types of deficiencies and their properties.\nthe physical functions of the dietary deficiencies are: - selenium (se) - zinc (zn) - copper (c) helps regulate the metabolism and metabolism of the human body. selenium is a component of the body's immune system\nthe image also includes a brief description of each type of deficiency and how they can affect the overall overall overall health of the individual.\nAtom = proton (p+) + neutron (n) + electron (e-) Atomic Number = no. of protons Mass Number = no. of neutrons Atomic Mass = mass of protons + mass of neutrons + mass of electrons Mass of\ndecaying isotopes can pose serious risks to living organisms. radioactive isotopes can be helpful, as in their use in dating fossils. living organisms cannot distinguish between isotopes.\nbiologists often use radioactive tracers to follow molecules as they undergo chemical changes in an organism. 14CO2 transforming into different metabolites during photosynthesis of carbohydrates.\nthe nucleus is represented by a pink circle with a plus sign in the center. on the right side, there is an electron cloud, which represents the nucleus of the atom. inside the nucleus, there are two electrons, one pink and\nimage suggests that the image is meant to illustrate the relationship between the two models. there is also a note on the image that reads \"note that these models are not to scale\"\nDNA, not protein contains the genetic code – proven by radioisotope labeling Griffith’s experiment (1928).\nafter infection by 32P-labeled phage, radioactivity was found only in the transduced bacteria. after infection with 35S-labeled phage, radioactivity is found only in the supernatant that remains after the bacteria are removed.\nin the first stage, there are six rats, each with a different color - R, S, and R. the rats are shown in a light brown color and are arranged in a grid-like pattern. the second stage is labelled \"S\",\nthe image shows the different types of heat treatments that can be used to treat the rats. the third stage is titled \"heat-treated\" and has a yellow sphere with a red arrow pointing towards it. the fourth stage is described as \"heat\ndiagram shows process of DNA, heat-treated S, and protease. diagram is divided into four sections, each representing a different stage of process. first section shows the DNA molecule, which is represented by a yellow circle.\nthe image shows a group of yellow spheres, which are likely the nucleotides of the molecule. the fourth section shows another brown mouse with a black arrow pointing towards the left side, representing the nucleotide.\nimage is a diagram that shows the process of centrifugation of bacteria in cells. it is divided into two sections. the first section on the left shows a sulfur-labeled protein capsule. the second section on the right shows a cell inside it\nthe capsule is labeled as \"bacteriophages\" and the cell inside the capsule is labeled as \"phosphorus labeled DNA (green)\" in the second section, there is an infection of the bacteria. the infection is represented by a green line that runs\nthe diagram shows how bacteria can be used to differentiate between different types of bacteria and how they interact with each other. the diagram also has a label that reads \"1\"\nsome radioactive isotopes are used for treatment. eg: the body uses iodine to make a thyroid hormone.\nradioactive iodine accumulates in the thyroid, so it can be used to kill thyroid cancer cells. substances that the body metabolizes, such as glucose or oxygen, may also be labeled. PET scan can produce images of areas of the body with\nAlzheimer is a neurodegenerative disease, wherein the brain becomes riddled with deposits (plaques) of a protein called beta-amyloid. researchers have synthesized a radioactively labeled protein molecule called PIB that binds\nimage shows two MRI images of the brain, one in blue and the other in yellow. the blue image shows the brain in detail, while the yellow image shows it in color. both images are labeled with the text \"healthy person\" and \"alzheimer\nuncontrolled exposure to high levels of radiation can be lethal. particles and energy thrown off by radioactive atoms can damage molecules, especially DNA.\nexplosion of nuclear reactor in Chernobyl, Ukraine, in 1986 released large amounts of radioactive isotopes into the environment. a few dozen people died from acute radiation poisoning within a few weeks of the accident. increased rates of thyroid cancer in children\nnatural sources of radiation can also pose a threat. radon is the second-leading cause of lung cancer in the united states. homeowners can buy a radon detector or hire a company to test their home.\ntechnology exists to remove radon from home if levels are found to be unsafe. Marie curie died of aplastic pernicious anemia due to her years of exposure to radiation during her work.\nthe diagram is divided into two columns, with the left column representing the element name and the right column representing half-life units. the elements are arranged in a grid-like pattern, with each element represented by a square box.\nthe atomic number is 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, and 92. distribution of electrons determine an atom's chemical properties.\nIonic bond: complete electron transfer between two atoms. covalent bond: sharing of electrons between two atoms.\nphotosynthesis: 6CO2 + 6H2O hvC6H12O6 + 6O2 Aerobic respiration: C6H12O6 + 6O2 6CO2 + 6H2O + ATP\nthe diagram is a flowchart that shows the process of photosynthesis. photosynthesis is a process where photosynthesis takes place in the atmosphere of a plant. the diagram shows a sun with a carbon dioxide molecule in the center, representing the energy of the\nphotosynthetic organisms use the energy to sunlight to manufacture glucose and other organic products, which heterotrophic cells use as energy and carbon sources. theory of spontaneous generation: if food was left to rot, it would generate living organisms like microbes, maggot\nDarwin’s theory of evolution suggested that simpler life forms gave rise to complex organisms. but pasteur said ‘life only comes from life’.\nabiogenic factors above in water gave rise to biomolecules (amino acids, lipids, nucleic acids, carbohydrates) all above theories were only speculations without proof.\nthe top section of the tree has a list of different types, including birds, reptiles, mammals, amphibians, land plants, and seaweed. there are also illustrations of animals such as a bird, a zebra, a ladybug, and\nseaweed is a type of seaweed that is native to the ocean floor. there is also an illustration that shows a snail, a crab, a jellyfish, a starfish, and an octopus. the image is labeled with\nsimulated the origins of life on earth by using – • water representing the oceans • reducing earth’s atmosphere consisting majorly of gases like CH4, NH3, H2 and H2O (water vapor) after a week, the water turned\namino acids, lipids and sugar are found on other planets, asteroids in our solar system and even beyond within our galaxy.\nit consists of a white sphere with a red arrow pointing towards it, representing the electrical spark. there are two pipes connected to the sphere, one on the left side and the other on the right side. the pipes are connected to each other by\nthe diagram appears to be a representation of the process of electrical spark lightning. spark lightning is a type of lightning that is used to generate electricity from a source of light.\nexperiment proved the Oparin-haldane chemical origin of life. free atoms of C, H, O, N • abiotic synthesis of monomers from simple inorganic gases like H2, CH4, NH3 and H2O.\nmonomers included amino acids, purines, pyrimidines, monosaccharides, fatty acids and glycerol. polymers include proteins, nucleic acids, lipids and polysaccharides.\nfirst cell probably used inorganic fuels produced by chemosynthesis for ATP generation: FeS + H2S ---FeS2 + H2 Figure 1-37, Landmarks in the evolution of life on earth lehninger.\nthe book is divided into six sections, each representing a different step in the process. the first section is titled \"Prebiotic formation of simple compounds, including nucleotides\" the second section is labeled \"Production of short RNA molecules with random sequences\nthe process is related to DNA genome translation and protein catalysts. a label at the bottom of the flowchart reads \"Genomic RNA begins to be copied into DNA\"\na table shows the different types of eukaryotes and their properties. the left column represents the number of different species of organisms. the right column represents their characteristics.\nthe characteristics include: 1,500, 2,000, and 3,000. there are also three types of photosynthetic O2 producing cyanobacteria. there is also a section titled \"Formation of oceans and continents\"\nfew aerobic bacteria evolved into mitochondria of modern eukaryotes. photosynthetic cyanobacteria became plastids, chloroplasts of green algae.\neach section represents a different type of chromosome. nuclei are responsible for storing nuclei in the nucleus of the cell. the nuclei are then released into the cell, which releases nuclei into the mitochondria.\nthe image also shows the mitochondrion, which acts as a catalyst for the cell's growth and development. in the center of the image, there are four chromosomes, each with a different color - blue, green, and white.\nthe white chromosomes represent chloroplasts, which are the cells that make up the cell membrane. the diagram also includes text that explains the process of photosynthesis and how photosynthesis can be used.\n","output_type":"stream"}]},{"cell_type":"code","source":"input_text=process_text_in_sentence_chunks(path,900)\nfor text in input_text:\n    print(generate_summary_bart(text))","metadata":{"execution":{"iopub.status.busy":"2024-06-27T18:02:31.464256Z","iopub.execute_input":"2024-06-27T18:02:31.464642Z","iopub.status.idle":"2024-06-27T18:03:40.672431Z","shell.execute_reply.started":"2024-06-27T18:02:31.464607Z","shell.execute_reply":"2024-06-27T18:03:40.671474Z"},"trusted":true},"execution_count":21,"outputs":[{"name":"stdout","text":"The logo is a circular emblem with a blue background and a white border. In the center of the emblem, there is an open book with a sunburst design around it.\nOrganisms are composed of elements, usually combined into compounds. Matter – occupies space and has mass. Three physical states of matter – solid, liquid and gas. All forms of matter are fundamentally composed of chemical elements.\nThe elements of the human body are C, O, H, N, Ca, P (make up 99% of body weight) Minor elements: K, S, Na, Cl, Mg (for nerve signaling and chemical reactions) Trace elements: B, Cr, Cu, F, I, Fe, Mn, Mo, Se, Si, Sn, V, Zn\nHe is standing with one leg slightly bent at the knee and the other leg extended behind him. The infographic is divided into three sections. The chart also includes information about the different elements and their properties.\nIodine deficiency leads to abnormal swelling of the thyroid gland leading to a condition called goiter. Iron deficiency anemia occurs due to lack of iron to produce hemoglobin. Iron rich foods include meat, fish, poultry, eggs, green, legumes, raisins, etc.\nExcess fluoride in drinking water can cause fluorosis, a condition of brown stained teeth. Excess/ overload of trace elements can lead to accidental fatal poisoning. Recommended dietary allowances (RDA) of trace Elements are based on gender (male/ female) and age.\nNext to the eggs, there is a pile of dried raisins. In the center of the table, there appears to be a large piece of raw meat, possibly pork or beef.\nThe table is divided into three columns and three rows. The first column is titled \"Physical Functions\" and lists the physical functions of the dietary deficiencies. There are also several other physical functions listed in the table, such as the body, the brain, and the brainstem.\nAtom means ‘indivisible’ but physicists split atom into sub-atomic particles. A radioactive isotope is one in which nucleus decays spontaneously. Radiation from decaying isotope can damage cellular molecules.\nLiving organisms cannot distinguish between isotopes. So, they uptake both normal and radioactive compounds similarly. Biologists often use radioactive tracers to follow molecules as they undergo chemical changes.\nInside the nucleus, there are two electrons, one pink and one black, representing protons. The protons are represented by the plus sign, while the neutrons represent the electrons. There is also a note on the image that reads \"Note that these models are not to scale\"\nDNA, not protein contains the genetic code – proven by radioisotope labeling. Viral DNA was labelled with 32P and protein with 35S. After E.coli infection by 32P-labeled phage, radioactivity was found only in the transduced bacteria.\nIn contrast, after infection with 35S-labeled phage, radioactivity is found only in the supernatant that remains after the bacteria are removed.\nThe diagram is divided into four sections, each representing a different stage of the process. The first section shows the DNA molecule, which is represented by a yellow circle with the letters \"DNA\" and \"heat-treated\" written on it.\nThe first section on the left shows a sulfur-labeled protein capsule with a cell inside it. The second section shows a group of yellow spheres, which are likely the nucleotides of the molecule. The fifth section shows that the molecule is protease, and the sixth section shows how it is treated with a heat-treated S.\nThe capsule is labeled as \"bacteriophages\" and the cell inside the capsule is labelled as \"phosphorus labeled DNA\" In the second section, there is an infection of the bacteria. The infection is represented by a green line that runs from the cell to the cell. The cell is shown in the center of the image.\nRadioactive isotopes may also be used to tag chemicals that accumulate in specific areas of the body. Substances that the body metabolizes, such as glucose or oxygen, may be labeled with a radioactive isotope. Some radioactive isotopes are used for treatment.\nPET (positron- emission tomography) scan can produce images of areas of the body with high metabolic activity. PET is useful for diagnosing certain heart disorders and cancers and for basic  research on the brain. Alzheimer is a neurodegenerative disease, wherein the brain becomes riddled with deposits (plaques) of a protein called beta-amyloid.\nBoth images are labeled with the text \"Healthy person\" and \"Alzheimer's patient\" The image also includes a text that reads \"Figure 2.4B PET images of brains of a healthy person (left) and a person with Alzheimer's disease (right) Red and yellow colors indicate high levels of PIB bound to beta-amyloid plaques\" This suggests\nRadon, a radioactive gas, is the second-leading cause of lung cancer in the U.S. Homeowners can buy a radon detector or hire a company to test their home to ensure that radon levels are safe.\nMarie Curie died of aplastic pernicious anemia (body stops producing blood cells) due to her years of exposure to radiation during her work.\nAt the top of the diagram, there is a title that reads \"The Uranium -238 Decay Chain\" below it, there are numbers that indicate the atomic decay chain. The atomic number is 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, and 92.\nEg: H2, O2, CH4,  H2O • Hydrogen bonds: weak bonds between polar molecules; important in the chemistry of life. Chemical reactions make and break chemical bonds 2H2 + O2 2H2O  Above reaction is an explosive reaction (fortunately does not occur in cells)\nThe diagram shows a sun with a carbon dioxide molecule in the center. The molecule is represented by a yellow circle with a red line connecting it to the carbon dioxide molecules. At the bottom of the diagram, there is a text that explains that photosynthesis is the ultimate source of all biological en-ergy.\nOparin-Haldane theory of abiogenesis (1929): Organic molecules needed for life could be formed from abiotic materials in the presence of energy. Oparin’s primordial soup: abiogenic factors above in water gave rise to biomolecules needed to form cells. All above theories were only speculations without proof.\nThere are also illustrations of animals such as a bird, a zebra, a ladybug, and a turtle. The animals are arranged in a circular pattern around the tree. There is also an illustration that shows a snail, a crab, a jellyfish, a starfish, and an octopus.\nMiller – Urey experiment (1953) Miller simulated the origins of life on earth by reducing earth’s atmosphere. After a week, the water turned brown and contained organic biomolecules. Amino acids, lipids and sugar are found on  other planets, asteroids in our solar system and even beyond within our galaxy.\nIn one set of experiments, Miller included H2S from atmospheric volcanoes, which accounted for formation of sulfur containing amino acids and subsequent disulfide links in proteins.\nOn the bottom left corner, there are two purple pipes, one of which is connected with a purple pipe. The red circle represents the light source, while the blue circle represents a blue sphere.\nMonomers included amino acids, purines, pyrimidines, monosaccharides, fatty acids and glycerol. Monomers combine to form polymers. Proteins, nucleic acids, lipids and polysaccharides form coacervates. Biomolecules are surrounded by a lipid membrane.\nFigure 1-36, A Possible RNA scenario world First cell probably used inorganic fuels produced by chemosynthesis for ATP generation. Figure 1-37, Landmarks in the evolution of life on earth.\nThe first section is titled \"Prebiotic formation of simple compounds, including nucleotides, from components of Earth's primitive atmosphere\" The second section is labeled \"Production of short RNA molecules with random sequences\" The third section is \"Selective replication of self-duplicating catalytic RNA segments\" The fourth section is labelled \"Synthesis of specific peptides, catalyzed\nThe table is divided into two columns, with the left column representing the number of different species of organisms. There are also three types of photosynthetic O2 producing cyanobacteria. At the bottom of the table, there is a section titled \"Formation of oceans and continents\"\nFew aerobic bacteria evolved into mitochondria of modern eukaryotes. Photosynthetic cyanobacteria became plastids, chloroplasts of green algae, the ancestors of modern plant cells.\nThe image also shows the mitochondrion, which acts as a catalyst for the cell's growth and development. In the center of the image, there are four chromosomes, each with a different color - blue, green, and white. These chromosomes are arranged in a circular pattern, representing the different stages of the cycle.\n","output_type":"stream"}]},{"cell_type":"code","source":"input_text=process_text_in_sentence_chunks(path,512)\nfor text in input_text:\n    print(generate_summary_pegasus(text))","metadata":{"execution":{"iopub.status.busy":"2024-06-27T18:08:23.606759Z","iopub.execute_input":"2024-06-27T18:08:23.607159Z","iopub.status.idle":"2024-06-27T18:15:12.853131Z","shell.execute_reply.started":"2024-06-27T18:08:23.607127Z","shell.execute_reply":"2024-06-27T18:15:12.852087Z"},"trusted":true},"execution_count":23,"outputs":[{"name":"stdout","text":"It is a circular emblem with a blue background and a white border.\nBelow the book, there are two lines of text in Hindi, which translates to \"Indian Institute of technology Indore\".\nOrganisms are composed of elements, usually combined into compounds Matter – occupies space and has mass Three physical states of matter – solid, liquid and gas All forms of matter – rocks, water, air and biological systems are fundamentally composed of chemical elements.\nEg: 2Na + Cl2 2 NaCl; (Both Na and Cl are corrosive, but NaCl is edible!)\n2H2 + O2 2H2O Compounds in living organisms comprise of 3 or 4 elements: • Carbohydrates: Carbon (C), Oxygen (O), Hydrogen (H) • Proteins: Carbon (C), Oxygen (O), Hydrogen (H), Nitrogen\nIt shows a silhouette of a man standing with his back to the viewer, facing the viewer.\nTrace elements are common additives to food and water Iodine is an essential component of the hormones thyroxine (T4 - inactive) and triiodothyronine (T3 - active) produced by the thyroid glands.\nIodine deficiency leads to abnormal swelling of the thyroid gland leading to a condition called goiter.\nIron deficiency anemia occurs due to lack of iron to produce hemoglobin. In pregnant women, iron deficiency puts the baby at the risk of developmental delays.\nRecommended dietary allowances (RDA) of trace elements are based on gender (male/ female) and age (infants/ adults) Image 1: Caption: The image is a portrait of an elderly woman from Burma.\nOn the left side of the image, there are two eggs, a bunch of green beans, and a small bowl of nuts.\nThe toothbrush has a green handle and white bristles, and the toothpaste is a light blue color.\nImportance of trace elements PN: Parenteral nutrition – feeding nutrients through intravenous/ non- gastrointestinal route Image 1: Caption: The image is a table that lists the different types of dietary deficiencies and their properties.\nThe physical functions are: - Selenium (S) - Zinc (Zn) As a component of the body's immune system, the body is responsible for regulating the flow of nutrients and nutrients in the body.\nThere are also several other physical functions listed in the table, such as the body, the brain, and the brainstem.\nof neutrons Atomic Mass = Mass of protons + Mass of neutrons + Mass of electrons Mass of proton = Mass of neutron  1 Da Mass of electron = 1/2000 Mass of proton; m (e-)  m (p+) So, Atomic Mass  Mass of\nRadiation from decaying isotope can damage cellular molecules and thus can pose serious risks to living organisms.\nThe model on the left is a nucleus, which is a type of atom with two protons and two neutrons.\nThe nucleus is represented by a pink circle with a plus sign in the center, representing the nucleus.\nThere is also a note on the image that reads \"Note that these models are not to scale; they greatly overestimate the size of the nucleus in relation to the electron cloud.\" This suggests that the image is meant to illustrate the relationship between the two models.\nDNA, not protein contains the genetic code – proven by radioisotope labeling Griffith’s experiment (1928) – bacterial transformation Two strains of the bacterium Streptococcus pneumoniae: • R-strain (rough colonies) – not virulent; did not kill mice • S-strain (\nIn contrast, after infection with 35S-labeled phage, radioactivity is found only in the supernatant that remains after the bacteria are removed.\nThe first stage is labeled \"R\" and shows a yellow circle with a black arrow pointing to it, indicating the direction of the heat treatment process.\nThe third stage is titled \"heat-treated\" and has a yellow sphere with a red arrow pointing towards it.\nThe first section shows the DNA molecule, which is represented by a yellow circle with the letters \"DNA\" and \"heat-treated\" written on it.\nThe fourth section shows another brown mouse with a black arrow pointing towards the left side, representing the nucleotide.\nThe image also shows the interaction between the two molecules, with the first molecule on the left and the second on the right.\nThe infection is represented by a green line that runs from the cell to the cell.\nThe arrows are pointing towards the cells, indicating that the infection is in the process.\nRadioactive isotopes in medical diagnosis and treatment Radioactive isotopes may also be used to tag chemicals that accumulate in specific areas of the body.\nPET (positron- emission tomography) scan can produce images of areas of the body with high metabolic activity.\nResearchers have synthesized a radioactively labeled protein molecule called PIB (Carbon-11- labeled Pittsburgh Compound B) that binds to beta-amyloid plaques and can be detected on a PET scan.\nImage 1: Caption: The image shows two MRI images of the brain, one in blue and the other in yellow. The blue image shows the brain in detail, while the yellow image shows it in color.\nThe particles and energy thrown off by radioactive atoms can damage molecules, especially DNA.\nIncreased rates of thyroid cancer in children exposed to the radiation have been reported.\nRadon, a radioactive gas, is the second-leading cause of lung cancer in the United States.\nImage 1: Caption: The image is a diagram of the Uranium-238 Decay Chain, which is a periodic table that shows the atomic number of the elements in the periodic table.\nThe second column is titled \"Element Names\" and lists the elements that are represented in the diagram.\nChemical bonds in biomolecules Distribution of electrons determine an atom’s chemical properties.\nEg: NaCl • Covalent bond: Sharing of electrons between two atoms.\nEg: H2, O2, CH4, H2O • Hydrogen bonds: weak bonds between polar molecules; important in the chemistry of life Chemical reactions make and break chemical bonds 2H2 + O2 2H2O Above reaction is an explosive reaction (fortunately does not occur\nThe diagram shows a sun with a carbon dioxide molecule in the center, representing the energy of the sun.\nTheory of spontaneous generation: If food was left to rot, it would generate living organisms like microbes, maggots (larvae of flies) and subsequently higher organisms like rats.\nOparin-Haldane theory of abiogenesis (1929): Organic molecules needed for life could be formed from abiotic materials in the presence of energy (lightning, heat and UV-light) under reducing conditions of early earth’s atmosphere (low levels of O2; major composition of\nImage 1: Caption: The image is a diagram of a tree with different types of plants and animals.\nThe animals are arranged in a circular pattern around the tree, with the birds at the top and the reptiles at the bottom.\nOn the right side, there is an illustration of a seaweed plant, which is a type of seaweed that is native to the ocean floor.\nMiller – Urey experiment (1953) Miller simulated the origins of life on earth by using – • Water representing the oceans • Reducing earth’s atmosphere consisting majorly of gases like CH4, NH3, H2 and H2O (water vapor) • Energy source – sunlight, geothermal\nPrebiotic chemistry – amino acids, lipids and sugar are found on other planets, asteroids in our solar system and even beyond within our galaxy.\nThere are two pipes connected to the sphere, one on the left side and the other on the right side.\nIn the center of the diagram, there is a red circle with a blue wave-like pattern, representing a light source.\nMiller – Urey experiment proved the Oparin-Haldane chemical origin of life Step1 • Free atoms of C, H, O, N • Abiotic synthesis of monomers from simple inorganic gases like H2, CH4, NH3 and H2O.\n• Monomers included amino acids, purines, pyrimidines, monosaccharides, fatty acids and glycerol Step2 • Monomers combine to form polymers • Proteins, nucleic acids, lipids and polysaccharides Step3 • Formation of coacervates • Biomolecules surrounded\nFigure 1-36, A Possible RNA scenario world First cell probably used inorganic fuels produced by chemosynthesis for ATP generation: FeS + H2S ---FeS2 + H2 Figure 1-37, Landmarks in the evolution of life on earth Lehninger, Principles of Biochemistry,\nThe second section is labeled \"Production of short RNA molecules with random sequences\".\nThe sixth section is called \"Primitive translation system develops, with RNA genome and RNA-protein catalysts\".\nThe table is divided into two columns, with the left column representing the number of different species of organisms and the right column representing their characteristics.\nThe appearance of protists, the first euaryotes, and the development of O2-rich atmosphere There are also three types of photosynthetic O2 producing cyanobacteria At the bottom of the table, there is a section titled \"Formation of oceans and continents\" which provides further\nImage 1: Caption: The image is a diagram that shows the different types of chromosomes and their functions.\nThe nuclei are responsible for storing nuclei in the nucleus of the cell.\nIn the center of the image, there are four chromosomes, each with a different color - blue, green, and white.\nThe diagram also includes text that explains the process of photosynthesis and how photosynthesis can be used to convert photosynthesis into photosynthesis.\n","output_type":"stream"}]},{"cell_type":"code","source":"","metadata":{"execution":{"iopub.status.busy":"2024-06-27T18:08:00.307123Z","iopub.status.idle":"2024-06-27T18:08:00.307510Z","shell.execute_reply.started":"2024-06-27T18:08:00.307337Z","shell.execute_reply":"2024-06-27T18:08:00.307352Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{"execution":{"iopub.status.busy":"2024-06-27T18:08:00.309172Z","iopub.status.idle":"2024-06-27T18:08:00.309571Z","shell.execute_reply.started":"2024-06-27T18:08:00.309363Z","shell.execute_reply":"2024-06-27T18:08:00.309377Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]}]}