{"metadata":{"kaggle":{"accelerator":"gpu","dataSources":[{"sourceId":8797460,"sourceType":"datasetVersion","datasetId":5289950},{"sourceType":"datasetVersion","sourceId":8908161,"datasetId":5356199}],"dockerImageVersionId":30733,"isInternetEnabled":true,"language":"python","sourceType":"notebook","isGpuEnabled":true},"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.13"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load\n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n\n# Input data files are available in the read-only \"../input/\" directory\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))\n\n# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session","metadata":{"_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","execution":{"iopub.status.busy":"2024-07-08T17:38:27.325186Z","iopub.execute_input":"2024-07-08T17:38:27.325892Z","iopub.status.idle":"2024-07-08T17:38:27.699288Z","shell.execute_reply.started":"2024-07-08T17:38:27.325861Z","shell.execute_reply":"2024-07-08T17:38:27.698433Z"},"trusted":true},"execution_count":3,"outputs":[{"name":"stdout","text":"/kaggle/input/bsebse/BSE102_Lecture 3_Chemical basis of life.pdf\n/kaggle/input/aaaaaaaaaaaaaaaaaaaaaaa/BSE102_Lecture 3_Chemical basis of life.pdf\n","output_type":"stream"}]},{"cell_type":"code","source":"pip install pymupdf","metadata":{"execution":{"iopub.status.busy":"2024-07-08T17:38:27.700817Z","iopub.execute_input":"2024-07-08T17:38:27.701206Z","iopub.status.idle":"2024-07-08T17:38:42.875598Z","shell.execute_reply.started":"2024-07-08T17:38:27.701180Z","shell.execute_reply":"2024-07-08T17:38:42.874461Z"},"trusted":true},"execution_count":4,"outputs":[{"name":"stdout","text":"Collecting pymupdf\n  Downloading PyMuPDF-1.24.7-cp310-none-manylinux2014_x86_64.whl.metadata (3.4 kB)\nCollecting PyMuPDFb==1.24.6 (from pymupdf)\n  Downloading PyMuPDFb-1.24.6-py3-none-manylinux2014_x86_64.manylinux_2_17_x86_64.whl.metadata (1.4 kB)\nDownloading PyMuPDF-1.24.7-cp310-none-manylinux2014_x86_64.whl (3.5 MB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.5/3.5 MB\u001b[0m \u001b[31m48.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n\u001b[?25hDownloading PyMuPDFb-1.24.6-py3-none-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (15.7 MB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m15.7/15.7 MB\u001b[0m \u001b[31m85.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m:00:01\u001b[0m00:01\u001b[0m\n\u001b[?25hInstalling collected packages: PyMuPDFb, pymupdf\nSuccessfully installed PyMuPDFb-1.24.6 pymupdf-1.24.7\nNote: you may need to restart the kernel to use updated packages.\n","output_type":"stream"}]},{"cell_type":"code","source":"pip install einops flash_attn","metadata":{"execution":{"iopub.status.busy":"2024-07-08T17:38:42.877572Z","iopub.execute_input":"2024-07-08T17:38:42.877942Z","iopub.status.idle":"2024-07-08T17:39:10.588810Z","shell.execute_reply.started":"2024-07-08T17:38:42.877908Z","shell.execute_reply":"2024-07-08T17:39:10.587746Z"},"trusted":true},"execution_count":5,"outputs":[{"name":"stdout","text":"Collecting einops\n  Downloading einops-0.8.0-py3-none-any.whl.metadata (12 kB)\nCollecting flash_attn\n  Downloading flash_attn-2.5.9.post1.tar.gz (2.6 MB)\n\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.6/2.6 MB\u001b[0m \u001b[31m34.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25ldone\n\u001b[?25hRequirement already satisfied: torch in /opt/conda/lib/python3.10/site-packages (from flash_attn) (2.1.2)\nRequirement already satisfied: filelock in /opt/conda/lib/python3.10/site-packages (from torch->flash_attn) (3.13.1)\nRequirement already satisfied: typing-extensions in /opt/conda/lib/python3.10/site-packages (from torch->flash_attn) (4.9.0)\nRequirement already satisfied: sympy in /opt/conda/lib/python3.10/site-packages (from torch->flash_attn) (1.12.1)\nRequirement already satisfied: networkx in /opt/conda/lib/python3.10/site-packages (from torch->flash_attn) (3.2.1)\nRequirement already satisfied: jinja2 in /opt/conda/lib/python3.10/site-packages (from torch->flash_attn) (3.1.2)\nRequirement already satisfied: fsspec in /opt/conda/lib/python3.10/site-packages (from torch->flash_attn) (2024.3.1)\nRequirement already satisfied: MarkupSafe>=2.0 in /opt/conda/lib/python3.10/site-packages (from jinja2->torch->flash_attn) (2.1.3)\nRequirement already satisfied: mpmath<1.4.0,>=1.1.0 in /opt/conda/lib/python3.10/site-packages (from sympy->torch->flash_attn) (1.3.0)\nDownloading einops-0.8.0-py3-none-any.whl (43 kB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m43.2/43.2 kB\u001b[0m \u001b[31m1.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hBuilding wheels for collected packages: flash_attn\n  Building wheel for flash_attn (setup.py) ... \u001b[?25ldone\n\u001b[?25h  Created wheel for flash_attn: filename=flash_attn-2.5.9.post1-cp310-cp310-linux_x86_64.whl size=120576656 sha256=0f3dddbf9bc350ea6b0306ec5ca5fee71b57fe1f06e6b72672690793d9dad2ce\n  Stored in directory: /root/.cache/pip/wheels/cc/ad/f6/7ccf0238790d6346e9fe622923a76ec218e890d356b9a2754a\nSuccessfully built flash_attn\nInstalling collected packages: einops, flash_attn\nSuccessfully installed einops-0.8.0 flash_attn-2.5.9.post1\nNote: you may need to restart the kernel to use updated packages.\n","output_type":"stream"}]},{"cell_type":"code","source":"pip install python-docx","metadata":{"execution":{"iopub.status.busy":"2024-07-08T17:39:10.591986Z","iopub.execute_input":"2024-07-08T17:39:10.592462Z","iopub.status.idle":"2024-07-08T17:39:23.292747Z","shell.execute_reply.started":"2024-07-08T17:39:10.592424Z","shell.execute_reply":"2024-07-08T17:39:23.291603Z"},"trusted":true},"execution_count":6,"outputs":[{"name":"stdout","text":"Collecting python-docx\n  Downloading python_docx-1.1.2-py3-none-any.whl.metadata (2.0 kB)\nRequirement already satisfied: lxml>=3.1.0 in /opt/conda/lib/python3.10/site-packages (from python-docx) (5.2.2)\nRequirement already satisfied: typing-extensions>=4.9.0 in /opt/conda/lib/python3.10/site-packages (from python-docx) (4.9.0)\nDownloading python_docx-1.1.2-py3-none-any.whl (244 kB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m244.3/244.3 kB\u001b[0m \u001b[31m5.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n\u001b[?25hInstalling collected packages: python-docx\nSuccessfully installed python-docx-1.1.2\nNote: you may need to restart the kernel to use updated packages.\n","output_type":"stream"}]},{"cell_type":"code","source":"pip install langchain","metadata":{"execution":{"iopub.status.busy":"2024-07-08T17:39:23.294252Z","iopub.execute_input":"2024-07-08T17:39:23.294561Z","iopub.status.idle":"2024-07-08T17:39:39.098435Z","shell.execute_reply.started":"2024-07-08T17:39:23.294535Z","shell.execute_reply":"2024-07-08T17:39:39.097330Z"},"trusted":true},"execution_count":7,"outputs":[{"name":"stdout","text":"Collecting langchain\n  Downloading langchain-0.2.6-py3-none-any.whl.metadata (7.0 kB)\nRequirement already satisfied: PyYAML>=5.3 in /opt/conda/lib/python3.10/site-packages (from langchain) (6.0.1)\nRequirement already satisfied: SQLAlchemy<3,>=1.4 in /opt/conda/lib/python3.10/site-packages (from langchain) (2.0.25)\nRequirement already satisfied: aiohttp<4.0.0,>=3.8.3 in /opt/conda/lib/python3.10/site-packages (from langchain) (3.9.1)\nRequirement already satisfied: async-timeout<5.0.0,>=4.0.0 in /opt/conda/lib/python3.10/site-packages (from langchain) (4.0.3)\nCollecting langchain-core<0.3.0,>=0.2.10 (from langchain)\n  Downloading langchain_core-0.2.11-py3-none-any.whl.metadata (6.0 kB)\nCollecting langchain-text-splitters<0.3.0,>=0.2.0 (from langchain)\n  Downloading langchain_text_splitters-0.2.2-py3-none-any.whl.metadata (2.1 kB)\nCollecting langsmith<0.2.0,>=0.1.17 (from langchain)\n  Downloading langsmith-0.1.84-py3-none-any.whl.metadata (13 kB)\nRequirement already satisfied: numpy<2,>=1 in /opt/conda/lib/python3.10/site-packages (from langchain) (1.26.4)\nRequirement already satisfied: pydantic<3,>=1 in /opt/conda/lib/python3.10/site-packages (from langchain) (2.5.3)\nRequirement already satisfied: requests<3,>=2 in /opt/conda/lib/python3.10/site-packages (from langchain) (2.32.3)\nRequirement already satisfied: tenacity!=8.4.0,<9.0.0,>=8.1.0 in /opt/conda/lib/python3.10/site-packages (from langchain) (8.2.3)\nRequirement already satisfied: attrs>=17.3.0 in /opt/conda/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (23.2.0)\nRequirement already satisfied: multidict<7.0,>=4.5 in /opt/conda/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (6.0.4)\nRequirement already satisfied: yarl<2.0,>=1.0 in /opt/conda/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (1.9.3)\nRequirement already satisfied: frozenlist>=1.1.1 in /opt/conda/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (1.4.1)\nRequirement already satisfied: aiosignal>=1.1.2 in /opt/conda/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (1.3.1)\nRequirement already satisfied: jsonpatch<2.0,>=1.33 in /opt/conda/lib/python3.10/site-packages (from langchain-core<0.3.0,>=0.2.10->langchain) (1.33)\nCollecting packaging<25,>=23.2 (from langchain-core<0.3.0,>=0.2.10->langchain)\n  Downloading packaging-24.1-py3-none-any.whl.metadata (3.2 kB)\nCollecting orjson<4.0.0,>=3.9.14 (from langsmith<0.2.0,>=0.1.17->langchain)\n  Downloading orjson-3.10.6-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (50 kB)\n\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m50.4/50.4 kB\u001b[0m \u001b[31m2.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hRequirement already satisfied: annotated-types>=0.4.0 in /opt/conda/lib/python3.10/site-packages (from pydantic<3,>=1->langchain) (0.6.0)\nRequirement already satisfied: pydantic-core==2.14.6 in /opt/conda/lib/python3.10/site-packages (from pydantic<3,>=1->langchain) (2.14.6)\nRequirement already satisfied: typing-extensions>=4.6.1 in /opt/conda/lib/python3.10/site-packages (from pydantic<3,>=1->langchain) (4.9.0)\nRequirement already satisfied: charset-normalizer<4,>=2 in /opt/conda/lib/python3.10/site-packages (from requests<3,>=2->langchain) (3.3.2)\nRequirement already satisfied: idna<4,>=2.5 in /opt/conda/lib/python3.10/site-packages (from requests<3,>=2->langchain) (3.6)\nRequirement already satisfied: urllib3<3,>=1.21.1 in /opt/conda/lib/python3.10/site-packages (from requests<3,>=2->langchain) (1.26.18)\nRequirement already satisfied: certifi>=2017.4.17 in /opt/conda/lib/python3.10/site-packages (from requests<3,>=2->langchain) (2024.2.2)\nRequirement already satisfied: greenlet!=0.4.17 in /opt/conda/lib/python3.10/site-packages (from SQLAlchemy<3,>=1.4->langchain) (3.0.3)\nRequirement already satisfied: jsonpointer>=1.9 in /opt/conda/lib/python3.10/site-packages (from jsonpatch<2.0,>=1.33->langchain-core<0.3.0,>=0.2.10->langchain) (2.4)\nDownloading langchain-0.2.6-py3-none-any.whl (975 kB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m975.5/975.5 kB\u001b[0m \u001b[31m24.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m\n\u001b[?25hDownloading langchain_core-0.2.11-py3-none-any.whl (337 kB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m337.4/337.4 kB\u001b[0m \u001b[31m19.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hDownloading langchain_text_splitters-0.2.2-py3-none-any.whl (25 kB)\nDownloading langsmith-0.1.84-py3-none-any.whl (127 kB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m127.9/127.9 kB\u001b[0m \u001b[31m8.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hDownloading orjson-3.10.6-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (141 kB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m141.1/141.1 kB\u001b[0m \u001b[31m9.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hDownloading packaging-24.1-py3-none-any.whl (53 kB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m54.0/54.0 kB\u001b[0m \u001b[31m3.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hInstalling collected packages: packaging, orjson, langsmith, langchain-core, langchain-text-splitters, langchain\n  Attempting uninstall: packaging\n    Found existing installation: packaging 21.3\n    Uninstalling packaging-21.3:\n      Successfully uninstalled packaging-21.3\n  Attempting uninstall: orjson\n    Found existing installation: orjson 3.9.10\n    Uninstalling orjson-3.9.10:\n      Successfully uninstalled orjson-3.9.10\n\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\ncudf 24.4.1 requires cubinlinker, which is not installed.\ncudf 24.4.1 requires cupy-cuda11x>=12.0.0, which is not installed.\ncudf 24.4.1 requires ptxcompiler, which is not installed.\ncuml 24.4.0 requires cupy-cuda11x>=12.0.0, which is not installed.\ndask-cudf 24.4.1 requires cupy-cuda11x>=12.0.0, which is not installed.\nkeras-cv 0.9.0 requires keras-core, which is not installed.\nkeras-nlp 0.12.1 requires keras-core, which is not installed.\ntensorflow-decision-forests 1.8.1 requires wurlitzer, which is not installed.\napache-beam 2.46.0 requires dill<0.3.2,>=0.3.1.1, but you have dill 0.3.8 which is incompatible.\napache-beam 2.46.0 requires numpy<1.25.0,>=1.14.3, but you have numpy 1.26.4 which is incompatible.\napache-beam 2.46.0 requires pyarrow<10.0.0,>=3.0.0, but you have pyarrow 14.0.2 which is incompatible.\ncudf 24.4.1 requires cuda-python<12.0a0,>=11.7.1, but you have cuda-python 12.5.0 which is incompatible.\ndistributed 2024.1.1 requires dask==2024.1.1, but you have dask 2024.5.2 which is incompatible.\ngoogle-cloud-bigquery 2.34.4 requires packaging<22.0dev,>=14.3, but you have packaging 24.1 which is incompatible.\njupyterlab 4.2.1 requires jupyter-lsp>=2.0.0, but you have jupyter-lsp 1.5.1 which is incompatible.\njupyterlab-lsp 5.1.0 requires jupyter-lsp>=2.0.0, but you have jupyter-lsp 1.5.1 which is incompatible.\nlibpysal 4.9.2 requires shapely>=2.0.1, but you have shapely 1.8.5.post1 which is incompatible.\nmomepy 0.7.0 requires shapely>=2, but you have shapely 1.8.5.post1 which is incompatible.\nosmnx 1.9.3 requires shapely>=2.0, but you have shapely 1.8.5.post1 which is incompatible.\nrapids-dask-dependency 24.4.1a0 requires dask==2024.1.1, but you have dask 2024.5.2 which is incompatible.\nrapids-dask-dependency 24.4.1a0 requires dask-expr==0.4.0, but you have dask-expr 1.1.2 which is incompatible.\nspopt 0.6.0 requires shapely>=2.0.1, but you have shapely 1.8.5.post1 which is incompatible.\ntensorflow 2.15.0 requires keras<2.16,>=2.15.0, but you have keras 3.3.3 which is incompatible.\nydata-profiling 4.6.4 requires numpy<1.26,>=1.16.0, but you have numpy 1.26.4 which is incompatible.\u001b[0m\u001b[31m\n\u001b[0mSuccessfully installed langchain-0.2.6 langchain-core-0.2.11 langchain-text-splitters-0.2.2 langsmith-0.1.84 orjson-3.10.6 packaging-24.1\nNote: you may need to restart the kernel to use updated packages.\n","output_type":"stream"}]},{"cell_type":"code","source":"pip install -U langchain-community","metadata":{"execution":{"iopub.status.busy":"2024-07-08T17:39:39.100019Z","iopub.execute_input":"2024-07-08T17:39:39.100606Z","iopub.status.idle":"2024-07-08T17:39:53.784550Z","shell.execute_reply.started":"2024-07-08T17:39:39.100557Z","shell.execute_reply":"2024-07-08T17:39:53.783444Z"},"trusted":true},"execution_count":8,"outputs":[{"name":"stdout","text":"Collecting langchain-community\n  Downloading langchain_community-0.2.6-py3-none-any.whl.metadata (2.5 kB)\nRequirement already satisfied: PyYAML>=5.3 in /opt/conda/lib/python3.10/site-packages (from langchain-community) (6.0.1)\nRequirement already satisfied: SQLAlchemy<3,>=1.4 in /opt/conda/lib/python3.10/site-packages (from langchain-community) (2.0.25)\nRequirement already satisfied: aiohttp<4.0.0,>=3.8.3 in /opt/conda/lib/python3.10/site-packages (from langchain-community) (3.9.1)\nRequirement already satisfied: dataclasses-json<0.7,>=0.5.7 in /opt/conda/lib/python3.10/site-packages (from langchain-community) (0.6.6)\nRequirement already satisfied: langchain<0.3.0,>=0.2.6 in /opt/conda/lib/python3.10/site-packages (from langchain-community) (0.2.6)\nRequirement already satisfied: langchain-core<0.3.0,>=0.2.10 in /opt/conda/lib/python3.10/site-packages (from langchain-community) (0.2.11)\nRequirement already satisfied: langsmith<0.2.0,>=0.1.0 in /opt/conda/lib/python3.10/site-packages (from langchain-community) (0.1.84)\nRequirement already satisfied: numpy<2,>=1 in /opt/conda/lib/python3.10/site-packages (from langchain-community) (1.26.4)\nRequirement already satisfied: requests<3,>=2 in /opt/conda/lib/python3.10/site-packages (from langchain-community) (2.32.3)\nRequirement already satisfied: tenacity!=8.4.0,<9.0.0,>=8.1.0 in /opt/conda/lib/python3.10/site-packages (from langchain-community) (8.2.3)\nRequirement already satisfied: attrs>=17.3.0 in /opt/conda/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain-community) (23.2.0)\nRequirement already satisfied: multidict<7.0,>=4.5 in /opt/conda/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain-community) (6.0.4)\nRequirement already satisfied: yarl<2.0,>=1.0 in /opt/conda/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain-community) (1.9.3)\nRequirement already satisfied: frozenlist>=1.1.1 in /opt/conda/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain-community) (1.4.1)\nRequirement already satisfied: aiosignal>=1.1.2 in /opt/conda/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain-community) (1.3.1)\nRequirement already satisfied: async-timeout<5.0,>=4.0 in /opt/conda/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain-community) (4.0.3)\nRequirement already satisfied: marshmallow<4.0.0,>=3.18.0 in /opt/conda/lib/python3.10/site-packages (from dataclasses-json<0.7,>=0.5.7->langchain-community) (3.21.2)\nRequirement already satisfied: typing-inspect<1,>=0.4.0 in /opt/conda/lib/python3.10/site-packages (from dataclasses-json<0.7,>=0.5.7->langchain-community) (0.9.0)\nRequirement already satisfied: langchain-text-splitters<0.3.0,>=0.2.0 in /opt/conda/lib/python3.10/site-packages (from langchain<0.3.0,>=0.2.6->langchain-community) (0.2.2)\nRequirement already satisfied: pydantic<3,>=1 in /opt/conda/lib/python3.10/site-packages (from langchain<0.3.0,>=0.2.6->langchain-community) (2.5.3)\nRequirement already satisfied: jsonpatch<2.0,>=1.33 in /opt/conda/lib/python3.10/site-packages (from langchain-core<0.3.0,>=0.2.10->langchain-community) (1.33)\nRequirement already satisfied: packaging<25,>=23.2 in /opt/conda/lib/python3.10/site-packages (from langchain-core<0.3.0,>=0.2.10->langchain-community) (24.1)\nRequirement already satisfied: orjson<4.0.0,>=3.9.14 in /opt/conda/lib/python3.10/site-packages (from langsmith<0.2.0,>=0.1.0->langchain-community) (3.10.6)\nRequirement already satisfied: charset-normalizer<4,>=2 in /opt/conda/lib/python3.10/site-packages (from requests<3,>=2->langchain-community) (3.3.2)\nRequirement already satisfied: idna<4,>=2.5 in /opt/conda/lib/python3.10/site-packages (from requests<3,>=2->langchain-community) (3.6)\nRequirement already satisfied: urllib3<3,>=1.21.1 in /opt/conda/lib/python3.10/site-packages (from requests<3,>=2->langchain-community) (1.26.18)\nRequirement already satisfied: certifi>=2017.4.17 in /opt/conda/lib/python3.10/site-packages (from requests<3,>=2->langchain-community) (2024.2.2)\nRequirement already satisfied: typing-extensions>=4.6.0 in /opt/conda/lib/python3.10/site-packages (from SQLAlchemy<3,>=1.4->langchain-community) (4.9.0)\nRequirement already satisfied: greenlet!=0.4.17 in /opt/conda/lib/python3.10/site-packages (from SQLAlchemy<3,>=1.4->langchain-community) (3.0.3)\nRequirement already satisfied: jsonpointer>=1.9 in /opt/conda/lib/python3.10/site-packages (from jsonpatch<2.0,>=1.33->langchain-core<0.3.0,>=0.2.10->langchain-community) (2.4)\nRequirement already satisfied: annotated-types>=0.4.0 in /opt/conda/lib/python3.10/site-packages (from pydantic<3,>=1->langchain<0.3.0,>=0.2.6->langchain-community) (0.6.0)\nRequirement already satisfied: pydantic-core==2.14.6 in /opt/conda/lib/python3.10/site-packages (from pydantic<3,>=1->langchain<0.3.0,>=0.2.6->langchain-community) (2.14.6)\nRequirement already satisfied: mypy-extensions>=0.3.0 in /opt/conda/lib/python3.10/site-packages (from typing-inspect<1,>=0.4.0->dataclasses-json<0.7,>=0.5.7->langchain-community) (1.0.0)\nDownloading langchain_community-0.2.6-py3-none-any.whl (2.2 MB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.2/2.2 MB\u001b[0m \u001b[31m30.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n\u001b[?25hInstalling collected packages: langchain-community\nSuccessfully installed langchain-community-0.2.6\nNote: you may need to restart the kernel to use updated packages.\n","output_type":"stream"}]},{"cell_type":"code","source":"# pip install flash_attn einops","metadata":{"execution":{"iopub.status.busy":"2024-07-08T17:39:53.786137Z","iopub.execute_input":"2024-07-08T17:39:53.786467Z","iopub.status.idle":"2024-07-08T17:39:53.791212Z","shell.execute_reply.started":"2024-07-08T17:39:53.786431Z","shell.execute_reply":"2024-07-08T17:39:53.790079Z"},"trusted":true},"execution_count":9,"outputs":[]},{"cell_type":"code","source":"from langchain.text_splitter import RecursiveCharacterTextSplitter\nfrom langchain.document_loaders import PyPDFLoader, DirectoryLoader\nfrom langchain.chains.summarize import load_summarize_chain\nfrom transformers import T5Tokenizer, T5ForConditionalGeneration\nfrom transformers import pipeline\nimport base64","metadata":{"execution":{"iopub.status.busy":"2024-07-08T17:39:53.792954Z","iopub.execute_input":"2024-07-08T17:39:53.793267Z","iopub.status.idle":"2024-07-08T17:40:10.169084Z","shell.execute_reply.started":"2024-07-08T17:39:53.793243Z","shell.execute_reply":"2024-07-08T17:40:10.168288Z"},"trusted":true},"execution_count":10,"outputs":[{"name":"stderr","text":"2024-07-08 17:40:00.165690: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n2024-07-08 17:40:00.165800: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n2024-07-08 17:40:00.299124: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n","output_type":"stream"}]},{"cell_type":"code","source":"import os\nimport fitz  \nfrom PIL import Image, ImageDraw, ImageFont\nimport io\nfrom io import BytesIO\nimport pytesseract  \nimport torch \nimport requests  \nfrom transformers import Blip2Processor, Blip2ForConditionalGeneration, PegasusForConditionalGeneration, PegasusTokenizer, AutoProcessor, AutoModelForCausalLM \nfrom docx import Document \nfrom docx.shared import Inches  \nimport tempfile  \nfrom transformers import T5Tokenizer, T5ForConditionalGeneration\nfrom transformers import BartForConditionalGeneration, BartTokenizer\nimport nltk\nfrom nltk.corpus import stopwords\nfrom nltk.tokenize import word_tokenize\nfrom transformers import BartTokenizer\nimport numpy as np \nimport pandas as pd \nnltk.download('punkt')\nnltk.download('stopwords')\n","metadata":{"execution":{"iopub.status.busy":"2024-07-08T17:40:10.170173Z","iopub.execute_input":"2024-07-08T17:40:10.170744Z","iopub.status.idle":"2024-07-08T17:40:11.345796Z","shell.execute_reply.started":"2024-07-08T17:40:10.170717Z","shell.execute_reply":"2024-07-08T17:40:11.344908Z"},"trusted":true},"execution_count":11,"outputs":[{"name":"stdout","text":"[nltk_data] Downloading package punkt to /usr/share/nltk_data...\n[nltk_data]   Package punkt is already up-to-date!\n[nltk_data] Downloading package stopwords to /usr/share/nltk_data...\n[nltk_data]   Package stopwords is already up-to-date!\n","output_type":"stream"},{"execution_count":11,"output_type":"execute_result","data":{"text/plain":"True"},"metadata":{}}]},{"cell_type":"code","source":"device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")","metadata":{"execution":{"iopub.status.busy":"2024-07-08T17:40:11.349218Z","iopub.execute_input":"2024-07-08T17:40:11.349810Z","iopub.status.idle":"2024-07-08T17:40:11.354205Z","shell.execute_reply.started":"2024-07-08T17:40:11.349784Z","shell.execute_reply":"2024-07-08T17:40:11.353316Z"},"trusted":true},"execution_count":12,"outputs":[]},{"cell_type":"code","source":"model = AutoModelForCausalLM.from_pretrained(\"microsoft/Florence-2-large\", trust_remote_code=True).to(device)\nprocessor = AutoProcessor.from_pretrained(\"microsoft/Florence-2-large\", trust_remote_code=True)","metadata":{"execution":{"iopub.status.busy":"2024-07-08T17:40:11.355226Z","iopub.execute_input":"2024-07-08T17:40:11.355556Z","iopub.status.idle":"2024-07-08T17:40:28.668765Z","shell.execute_reply.started":"2024-07-08T17:40:11.355520Z","shell.execute_reply":"2024-07-08T17:40:28.667748Z"},"trusted":true},"execution_count":13,"outputs":[{"output_type":"display_data","data":{"text/plain":"config.json:   0%|          | 0.00/2.44k [00:00<?, ?B/s]","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"42ce7321d28842e8be29d3bf4196743d"}},"metadata":{}},{"output_type":"display_data","data":{"text/plain":"configuration_florence2.py:   0%|          | 0.00/15.1k [00:00<?, ?B/s]","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"6fba47b4c0be43668645dbc071e6b172"}},"metadata":{}},{"name":"stderr","text":"A new version of the following files was downloaded from https://huggingface.co/microsoft/Florence-2-large:\n- configuration_florence2.py\n. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.\n","output_type":"stream"},{"output_type":"display_data","data":{"text/plain":"modeling_florence2.py:   0%|          | 0.00/127k [00:00<?, ?B/s]","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"a197c4e1892e42e385fdb1289a78a664"}},"metadata":{}},{"name":"stderr","text":"A new version of the following files was downloaded from https://huggingface.co/microsoft/Florence-2-large:\n- modeling_florence2.py\n. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.\n","output_type":"stream"},{"output_type":"display_data","data":{"text/plain":"pytorch_model.bin:   0%|          | 0.00/1.54G [00:00<?, ?B/s]","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"5c83c13f82ca4fedab2f0a515226cf07"}},"metadata":{}},{"name":"stderr","text":"/opt/conda/lib/python3.10/site-packages/torch/_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()\n  return self.fget.__get__(instance, owner)()\n","output_type":"stream"},{"output_type":"display_data","data":{"text/plain":"generation_config.json:   0%|          | 0.00/51.0 [00:00<?, ?B/s]","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"fef7977d42c241f3858bc7059152f44c"}},"metadata":{}},{"output_type":"display_data","data":{"text/plain":"preprocessor_config.json:   0%|          | 0.00/806 [00:00<?, ?B/s]","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"83c806d425da42ae80b3afc81aca34ba"}},"metadata":{}},{"output_type":"display_data","data":{"text/plain":"processing_florence2.py:   0%|          | 0.00/46.4k [00:00<?, ?B/s]","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"9c437ed349c04342bba60e23d125e12d"}},"metadata":{}},{"name":"stderr","text":"A new version of the following files was downloaded from https://huggingface.co/microsoft/Florence-2-large:\n- processing_florence2.py\n. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.\n","output_type":"stream"},{"output_type":"display_data","data":{"text/plain":"tokenizer_config.json:   0%|          | 0.00/34.0 [00:00<?, ?B/s]","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"f828d710be73418ea61a02c3d0e8a56a"}},"metadata":{}},{"output_type":"display_data","data":{"text/plain":"vocab.json:   0%|          | 0.00/1.10M [00:00<?, ?B/s]","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"db49f0697f0b4053a1100622bd197682"}},"metadata":{}},{"output_type":"display_data","data":{"text/plain":"tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"d4c9f6f266a14157a742bfffe20ceda3"}},"metadata":{}}]},{"cell_type":"code","source":"def run_example(image, task_prompt, text_input=None):\n    if text_input is None:\n        prompt = task_prompt\n    else:\n        prompt = task_prompt + text_input\n    inputs = processor(text=prompt, images=image, return_tensors=\"pt\").to(device)\n    generated_ids = model.generate(\n        input_ids=inputs[\"input_ids\"],\n        pixel_values=inputs[\"pixel_values\"],\n        max_new_tokens=1024,\n        num_beams=3\n    )\n    generated_text = processor.batch_decode(generated_ids, skip_special_tokens=False)[0]\n    parsed_answer = processor.post_process_generation(generated_text, task=task_prompt, image_size=(image.width, image.height))\n    caption = parsed_answer[\"<MORE_DETAILED_CAPTION>\"]\n    return caption","metadata":{"execution":{"iopub.status.busy":"2024-07-08T17:40:28.670155Z","iopub.execute_input":"2024-07-08T17:40:28.670662Z","iopub.status.idle":"2024-07-08T17:40:28.677970Z","shell.execute_reply.started":"2024-07-08T17:40:28.670626Z","shell.execute_reply":"2024-07-08T17:40:28.677051Z"},"trusted":true},"execution_count":14,"outputs":[]},{"cell_type":"code","source":"def pdf_doc(pdf_document):\n    doc = fitz.open(pdf_document)\n\n    page_text_list = []\n    for page_number in range(len(doc)):\n        page = doc.load_page(page_number)  \n        image_list = page.get_images(full=True)  \n\n        page_text = page.get_text()\n        page_text_combined = page_text\n\n        \n        for img_index, img in enumerate(image_list):\n            xref = img[0] \n            base_image = doc.extract_image(xref)  \n            image_bytes = base_image[\"image\"]\n            \n            image = Image.open(io.BytesIO(image_bytes)).convert('RGB')\n            \n            prompt = \"<MORE_DETAILED_CAPTION>\"\n            generated_caption = run_example(image, prompt)\n\n            combined_text = f\"\\n\\nImage {img_index + 1}:\\nCaption: {generated_caption}\\n\"\n\n            page_text_combined += combined_text\n\n        page_text_list.append(page_text_combined)\n    return page_text_list","metadata":{"execution":{"iopub.status.busy":"2024-07-08T17:40:28.679253Z","iopub.execute_input":"2024-07-08T17:40:28.679599Z","iopub.status.idle":"2024-07-08T17:40:32.447001Z","shell.execute_reply.started":"2024-07-08T17:40:28.679568Z","shell.execute_reply":"2024-07-08T17:40:32.445966Z"},"trusted":true},"execution_count":15,"outputs":[]},{"cell_type":"code","source":"def docx_doc(docx_document):\n    doc = Document(docx_document)\n    doc_text_list = []\n    for para in doc.paragraphs:\n        doc_text_list.append(para.text)\n\n        for run in para.runs:\n            if run.element.xml.find(\"pic:blipFill\") != -1:\n                image = run.element.xpath(\".//a:blip\")[0]\n                image_data = image.get(\"{http://schemas.openxmlformats.org/officeDocument/2006/relationships}embed\")\n                image_part = doc.part.related_parts[image_data]\n                image_bytes = image_part.blob\n\n                image = Image.open(io.BytesIO(image_bytes)).convert('RGB')\n                prompt = \"<MORE_DETAILED_CAPTION>\"\n                generated_caption = run_example(image, prompt)\n                combined_text = f\"Caption: {generated_caption}\\n\"\n                doc_text_list.append(combined_text)\n    return doc_text_list\n\n    \n\n","metadata":{"execution":{"iopub.status.busy":"2024-07-08T17:40:32.448197Z","iopub.execute_input":"2024-07-08T17:40:32.448566Z","iopub.status.idle":"2024-07-08T17:40:32.462032Z","shell.execute_reply.started":"2024-07-08T17:40:32.448539Z","shell.execute_reply":"2024-07-08T17:40:32.461236Z"},"trusted":true},"execution_count":16,"outputs":[]},{"cell_type":"code","source":"def determine_file_type(file_path):\n    _, file_extension = os.path.splitext(file_path)\n    if file_extension.lower() == '.docx':\n        return 'docx'\n    elif file_extension.lower() == '.pdf':\n        return 'pdf'\n    else:\n        return 'unknown'","metadata":{"execution":{"iopub.status.busy":"2024-07-08T17:40:32.463077Z","iopub.execute_input":"2024-07-08T17:40:32.463340Z","iopub.status.idle":"2024-07-08T17:40:32.473496Z","shell.execute_reply.started":"2024-07-08T17:40:32.463316Z","shell.execute_reply":"2024-07-08T17:40:32.472753Z"},"trusted":true},"execution_count":17,"outputs":[]},{"cell_type":"code","source":"def get_text(path):\n    text_each_page=[]\n    file_type=determine_file_type(path)\n    if file_type=='pdf':\n        text_each_page=pdf_doc(path)\n    elif file_type=='docx':\n        text_each_page=docx_doc(path)\n    else :\n        print(\"invalid file type\")\n    return text_each_page","metadata":{"execution":{"iopub.status.busy":"2024-07-08T17:40:32.474525Z","iopub.execute_input":"2024-07-08T17:40:32.474806Z","iopub.status.idle":"2024-07-08T17:40:32.483619Z","shell.execute_reply.started":"2024-07-08T17:40:32.474783Z","shell.execute_reply":"2024-07-08T17:40:32.482773Z"},"trusted":true},"execution_count":18,"outputs":[]},{"cell_type":"code","source":"def preprocess_text(text):\n    text = text.lower()\n    tokens = word_tokenize(text)\n    stop_words = set(stopwords.words('english'))\n    tokens = [word for word in tokens if word.isalpha() and word not in stop_words]\n    return ' '.join(tokens)\n","metadata":{"execution":{"iopub.status.busy":"2024-07-08T17:40:32.484731Z","iopub.execute_input":"2024-07-08T17:40:32.485004Z","iopub.status.idle":"2024-07-08T17:40:32.498306Z","shell.execute_reply.started":"2024-07-08T17:40:32.484981Z","shell.execute_reply":"2024-07-08T17:40:32.497458Z"},"trusted":true},"execution_count":19,"outputs":[]},{"cell_type":"code","source":"path=\"/kaggle/input/aaaaaaaaaaaaaaaaaaaaaaa/BSE102_Lecture 3_Chemical basis of life.pdf\"\n","metadata":{"execution":{"iopub.status.busy":"2024-07-08T17:45:39.256045Z","iopub.status.idle":"2024-07-08T17:45:39.256439Z","shell.execute_reply.started":"2024-07-08T17:45:39.256234Z","shell.execute_reply":"2024-07-08T17:45:39.256249Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# def process_text_in_sentence_chunks(path, max_chunk_size):\n#     text=('\\n'.join(page for page in get_text(path)))\n#     sentences = text.replace(\"\\n\", \" \") \n#     sentences = nltk.sent_tokenize(sentences)\n\n#     current_chunk = \"\"\n#     text_list = []\n#     for sentence in sentences:\n#         if len(current_chunk) + len(sentence) <= max_chunk_size:\n#             current_chunk += \" \" + sentence\n#         else:\n#               text_list.append(current_chunk.strip())\n#               current_chunk = sentence\n\n#     if current_chunk:\n#         text_list.append(current_chunk.strip())\n#     return text_list","metadata":{"execution":{"iopub.status.busy":"2024-07-08T17:41:20.734588Z","iopub.execute_input":"2024-07-08T17:41:20.734883Z","iopub.status.idle":"2024-07-08T17:41:20.742131Z","shell.execute_reply.started":"2024-07-08T17:41:20.734858Z","shell.execute_reply":"2024-07-08T17:41:20.741226Z"},"trusted":true},"execution_count":21,"outputs":[]},{"cell_type":"code","source":"def file_preprocessing(file):\n    pages=('\\n'.join(page for page in get_text(path)))\n    text_splitter = RecursiveCharacterTextSplitter(chunk_size=200, chunk_overlap=50)\n    texts = text_splitter.split_text(pages)\n    final_texts = \"\"\n    for text in texts:\n        final_texts = final_texts + text\n    return final_texts","metadata":{"execution":{"iopub.status.busy":"2024-07-08T18:10:26.190472Z","iopub.execute_input":"2024-07-08T18:10:26.191292Z","iopub.status.idle":"2024-07-08T18:10:26.196616Z","shell.execute_reply.started":"2024-07-08T18:10:26.191260Z","shell.execute_reply":"2024-07-08T18:10:26.195581Z"},"trusted":true},"execution_count":51,"outputs":[]},{"cell_type":"code","source":"k=file_preprocessing(path)","metadata":{"execution":{"iopub.status.busy":"2024-07-08T18:10:29.241626Z","iopub.execute_input":"2024-07-08T18:10:29.241986Z","iopub.status.idle":"2024-07-08T18:11:16.570820Z","shell.execute_reply.started":"2024-07-08T18:10:29.241956Z","shell.execute_reply":"2024-07-08T18:11:16.569829Z"},"trusted":true},"execution_count":52,"outputs":[]},{"cell_type":"code","source":"print(k)","metadata":{"execution":{"iopub.status.busy":"2024-07-08T18:11:38.413062Z","iopub.execute_input":"2024-07-08T18:11:38.413727Z","iopub.status.idle":"2024-07-08T18:11:38.419564Z","shell.execute_reply.started":"2024-07-08T18:11:38.413697Z","shell.execute_reply":"2024-07-08T18:11:38.418671Z"},"trusted":true},"execution_count":53,"outputs":[{"name":"stdout","text":"BSE 102 Bioscience\nSpring 2024\nThe Chemical Basis of Life\nInstructor: Dr. Sunil Kumar Boda\nAssistant Professor\nDepartment of Biosciences & Biomedical Engineering\nIIT IndoreIIT Indore\nE-mail: sunilboda@iiti.ac.inImage 1:\nCaption:The image is the logo of the Indian Institute of Technology Indore. It is a circular emblem with a blue background and a white border. In the center of the emblem, there is an open book with aof the emblem, there is an open book with a sunburst design around it. The book is surrounded by a banner with the words \"IIT Indore\" written in Hindi. Below the book, there are two lines of text inBelow the book, there are two lines of text in Hindi, which translates to \"Indian Institute of technology Indore\". The text below the book reads \"II संस्थान सेवारीय साथिकाली\" (Indian Institute forसंस्थान सेवारीय साथिकाली\" (Indian Institute for Technology and Technology) in English. The emblem is a symbol of knowledge and knowledge in India.Organisms are composed of elements, usually combined into compounds\nMatter – occupies space and has mass\nThree physical states of matter – solid, liquid and gasAll forms of matter – rocks, water, air and biological systems are fundamentally composed \nof chemical elements.\nAn element is a substance that cannot be further broken down to other substances byordinary chemical means. \nA compound is a substance containing two or more elements combined in a fixed ratio. \nEg: 2Na + Cl2 2 NaCl; (Both Na and Cl are corrosive, but NaCl is edible!)2H2 + O2 2H2O\nCompounds in living organisms comprise of 3 or 4 elements:\n•\nCarbohydrates: Carbon (C), Oxygen (O), Hydrogen (H)\n••\nProteins: Carbon (C), Oxygen (O), Hydrogen (H), Nitrogen (N), Sulfur (S)\n•\nNucleic acids: Carbon (C), Oxygen (O), Hydrogen (H), Nitrogen (N), Phosphorus (P)\n••\nLipids: Carbon (C), Oxygen (O), Hydrogen (H), Phosphorus (P)\nMajor elements in the human body: C, O, H, N, Ca, P (make up 99% of body weight)Minor elements (< 1 wt%): K, S, Na, Cl, Mg (for nerve signaling and chemical reactions)\nTrace elements (< 0.01 wt%): B, Cr, Cu, F, I, Fe, Mn, Mo, Se, Si, Sn, V, ZnImage 1:\nCaption:The image is an infographic that shows the elements of the human body as per-centage of body weight (including water). It shows a silhouette of a man standing with his back to the viewer, facing thestanding with his back to the viewer, facing the viewer. The man is wearing a red t-shirt, black shorts, and colorful socks. He is standing with one leg slightly bent at the knee and the other legleg slightly bent at the knee and the other leg extended behind him.The infographic is divided into three sections. The first section is titled \"Figure 2.1B\" and shows the percentage of water in each element. The second section is labeled \"Oxygen (O): 65%\". The thirdsection is labeled \"Oxygen (O): 65%\". The third section is labelled \"Carbon (C): 18.5%\", the fourth section is \"Hydrogen (H): 9.5%, the fifth section is listed as \"Nitrogen (N): 3.3%\", and the sixthis listed as \"Nitrogen (N): 3.3%\", and the sixth section is colored in shades of blue, green, yellow, and orange. The chart also includes information about the different elements and theirabout the different elements and their properties.Trace elements are common additives to food and water\nIodine is an essential component of the hormones thyroxine (T4 - inactive) and triiodothyronine (T3 - active)produced by the thyroid glands. Thyroid hormones control body’s metabolism (food to energy conversion) rate.Iodine deficiency leads to abnormal swelling of the thyroid gland leading to a condition called goiter. Mostserious effects of iodine deficiency occur during fetal development and childhood, leading to miscarriages, \npoor growth and mental impairment.poor growth and mental impairment. \nIodine sources include seafood, kelp, dairy products and dark leafy greens. Iodized salt contains 45 μg/g or \nppm of table salt. Daily requirement is 150 μg.ppm of table salt. Daily requirement is 150 μg.\nIron (Fe) makes up only 0.004 wt% of body weight, but it is vital for energy \nprocessing and transport of oxygen.processing and transport of oxygen. \nIron deficiency anemia occurs due to lack of iron to produce hemoglobin. In \npregnant women, iron deficiency puts the baby at the risk of developmental delays.Iron rich foods include meat, fish, poultry, eggs, green, legumes, raisins, etc.\nFluoride is added to toothpaste and mouthwash at ~1500 ppm to prevent dental cavities and tooth loss.Excess fluoride in drinking water can cause fluorosis, a condition of brown stained teeth.\nExcess/ overload of trace elements can lead to accidental fatal poisoning. Recommended dietaryallowances (RDA) of trace elements are based on gender (male/ female) and age (infants/ adults)Image 1:\nCaption:The image is a portrait of an elderly woman from Burma. She is wearing a blue and white headscarf with a pink scarf around her neck. The woman has a serious expression on her face and is lookinga serious expression on her face and is looking directly at the camera. She appears to be in her late 60s or early 70s. The background is plain white. The text on the image reads \"Figure 2.2A Goiter,The text on the image reads \"Figure 2.2A Goiter, a symptom of iodine deficiency, in a Burma woman.\"Image 2:\nCaption:The image shows a variety of food items that are rich in iron. On the left side of the image, there are two eggs, a bunch of green beans, and a small bowl of nuts. Next to the eggs, there is a pilebowl of nuts. Next to the eggs, there is a pile of dried raisins. In the center of the table, there appears to be a large piece of raw meat, possibly pork or beef, with a pink onion and parsley onpork or beef, with a pink onion and parsley on top. The meat is pink in color and looks to be seasoned with herbs and spices. The background is white.Image 3:\nCaption:The image shows a toothbrush and a glass of toothpaste. The toothbrush has a green handle and white bristles, and the toothpaste is a light blue color. The glass is filled with a clear liquid, whichThe glass is filled with a clear liquid, which appears to be toothpaste with added fluoride. The background is white, and there is text on the image that reads \"Figure 2.2C Mouthwash and toothpaste\".Importance of trace elements\nPN: Parenteral nutrition –\nfeeding nutrients through \nintravenous/ non-\ngastrointestinal routeImage 1:\nCaption: \nThe image is a table that lists the different types of dietary deficiencies and their properties. The table is divided into three columns and three rows.The first column is titled \"Physical Functions\" and lists the physical functions of the dietary deficiencies. The physical functions are:\n\n- Selenium (S)\n- Zinc (Zn)As a component of the body's immune system, the body is responsible for regulating the flow of nutrients and nutrients in the body.- Copper (C) is a component that helps to regulate the metabolism and metabolism of the human body. It helps to reduce inflammation and improve the overall health and wellbeing of the person.There are also several other physical functions listed in the table, such as the body, the brain, and the brainstem. These functions are essential for maintaining a healthy and well-being, and canfor maintaining a healthy and well-being, and can help to maintain a balanced diet and maintain a healthy lifestyle. The image also includes a brief description of each type of deficiency and howdescription of each type of deficiency and how they can affect the overall overall health of the individual.Radioactive isotopes in biology\nAtom means ‘indivisible’. But physicists split atom into sub-atomic particles.\nAtom = proton (p+) + neutron (n) + electron (e-)\nAtomic Number = No. of protonsAtomic Number = No. of protons\nMass Number = No. of protons + No. of neutrons\nAtomic Mass = Mass of protons + Mass of neutrons + Mass of electrons\nMass of proton = Mass of neutron ~ 1 DaMass of proton = Mass of neutron ~ 1 Da\nMass of electron = 1/2000 Mass of proton; m (e-) << m (p+)\nSo, Atomic Mass ≈ Mass of protons + Mass of neutronsIsotopes = same atomic number, but different mass number. Eg: 6\n12𝐶,\n6\n13𝐶,\n6\n14𝐶\nA radioactive isotope is one in which nucleus decays spontaneously, emitting particles and energy.Radiation from decaying isotope can damage cellular molecules and thus can pose serious risks to living organisms.But radioactive isotopes can be helpful, as in their use in dating fossils by C-14 decay (t1/2 = 5730 years).Living organisms cannot distinguish between isotopes. So, they uptake both normal and radioactive compounds similarly.Basic research involving radioisotopes: Biologists often use radioactive tracers to follow molecules as they undergo chemicalchanges in an organism. Eg: Monitoring 14CO2 transforming into different metabolites during photosynthesis of carbohydrates\nAtomic structureImage 1:\nCaption:The image is a diagram that shows two models of a helium atom. The model on the left is a nucleus, which is a type of atom with two protons and two neutrons. The nucleus is represented by a pinkneutrons. The nucleus is represented by a pink circle with a plus sign in the center, representing the nucleus.On the right side of the image, there is an electron cloud, which represents the nucleus of the atom. Inside the nucleus, there are two electrons, one pink and one black, representing protons. Theone pink and one black, representing protons. The protons are represented by the plus sign, while the neutrons represent the electrons.There is also a note on the image that reads \"Note that these models are not to scale; they greatly overestimate the size of the nucleus in relation to the electron cloud.\" This suggests that theto the electron cloud.\" This suggests that the image is meant to illustrate the relationship between the two models.DNA, not protein contains the genetic code – proven by radioisotope labeling \nGriffith’s experiment (1928) – bacterial transformation\nTwo strains of the bacterium Streptococcus pneumoniae:\n••\nR-strain (rough colonies) – not virulent; did not kill mice\n•\nS-strain (smooth colonies) – virulent; killed infected mice\n•\nHeat-treated S-strain – non-virulent; did not kill mice\n••\nHeat-treated S-strain + R-strain – killed mice; horizontal gene \ntransfer from heat-treated S transformed R-strain to virulent \nstrain\nHershey & Chase experiment (1952) - transductionHershey & Chase experiment (1952) - transduction\nViral DNA was labelled with 32P and protein with 35S. After \nE.coli infection by 32P-labeled phage, radioactivity was foundonly in the transduced bacteria (after the phage are removed \nby agitation and centrifugation). In contrast, after infection \nwith 35S-labeled phage, radioactivity is found only in thesupernatant that remains after the bacteria are removed.Image 1:\nCaption:Image 1:\nCaption: \nThe image is a diagram that shows the process of heat-treated S and S+R. It shows the different stages of the process, from the initial stage of heat treatment to the final stage.In the first stage, there are six rats, each with a different color - R, S, and R. The rats are shown in a light brown color and are arranged in a grid-like pattern. The first stage is labeled \"R\"grid-like pattern. The first stage is labeled \"R\" and shows a yellow circle with a black arrow pointing to it, indicating the direction of the heat treatment process. The second stage is labelledtreatment process. The second stage is labelled \"S\", indicating that the rats are being treated with a heat treatment. The third stage is titled \"heat-treated\" and has a yellow sphere with a red\"heat-treated\" and has a yellow sphere with a red arrow pointing towards it. The fourth stage is described as \"heat treated S\" and is labeled as \"R\".There are also several smaller yellow spheres scattered throughout the image, representing the different types of heat treatments that can be used to treat the rats. These are likely to be used into treat the rats. These are likely to be used in the treatment process, such as heat treatment, treatment, and treatment.Image 2:\nCaption: \nThe image is a diagram that shows the process of DNA, heat-treated S, and protease. The diagram is divided into four sections, each representing a different stage of the process.The first section shows the DNA molecule, which is represented by a yellow circle with the letters \"DNA\" and \"heat-treated\" written on it. The molecule is surrounded by a black circle with a yellowis surrounded by a black circle with a yellow arrow pointing to it, indicating that it is being treated with the DNA. The second section shows a group of yellow spheres, which are likely thea group of yellow spheres, which are likely the nucleotides of the molecule. The third section shows two brown mice, one on each side of the image, with their heads facing towards the right side. Thetheir heads facing towards the right side. The fourth section shows another brown mouse with a black arrow pointing towards the left side, representing the nucleotide. The fifth section shows thatthe nucleotide. The fifth section shows that the molecule is protease, and the sixth section shows how it is treated with a heat-treated S. The image also shows the interaction between the twoimage also shows the interaction between the two molecules, with the first molecule on the left and the second on the right.Image 3:\nCaption: \nThe image is a diagram that shows the process of centrifugation of bacteria in cells. It is divided into two sections.The first section on the left shows a sulfur-labeled protein capsule (red) with a cell inside it. The capsule is labeled as \"bacteriophages\" and the cell inside the capsule is labelled as \"phosphorusinside the capsule is labelled as \"phosphorus labeled DNA (green).In the second section, there is an infection of the bacteria. The infection is represented by a green line that runs from the cell to the cell. The cell is shown in the center of the image.There are three steps in the diagram:\n\n1. Infection\n2. Blending\n3. Centrifugation\n4. After centrifugating no sulfur in cellsAt the bottom of the diagram, there are two circles, one labeled \"cell\" and another labeled \"centrifugal\". The circles are connected by arrows, indicating the direction of the infection. The arrowsthe direction of the infection. The arrows are pointing towards the cells, indicating that the infection is in the process. The diagram also has a label that reads \"1.\"Overall, the image shows how bacteria can be used to differentiate between different types of bacteria and how they interact with each other.Radioactive isotopes in medical diagnosis and treatment\nRadioactive isotopes may also be used to tag chemicals that accumulate in specific areas of the body. Eg: Phosphorus (30P) in \nbones.bones. \n[18F-FDG]Fluorodeoxyglucose is used to detect cancer. After injection of tiny amount of such a tracer, a special camera \nproduces an image of where the radiation collects.Some radioactive isotopes are used for treatment. Eg: The body uses iodine to make a thyroid hormone. Because radioactiveiodine accumulates in the thyroid, it can be used to kill thyroid cancer cells.Substances that the body metabolizes, such as glucose or oxygen, may also be labeled with a radioactive isotope. PET (positron-emission tomography) scan can produce images of areas of the body with high metabolic activity.\nPET is useful for diagnosing certain heart disorders and cancers and for basic \nresearch on the brain.research on the brain.\nAlzheimer is a neurodegenerative disease, wherein the brain becomes riddled \nwith deposits (plaques) of a protein called beta-amyloid. Researchers havesynthesized a radioactively labeled protein molecule called PIB (Carbon-11-\nlabeled Pittsburgh Compound B) that binds to beta-amyloid plaques and can be \ndetected on a PET scan.detected on a PET scan.\nBrain of the Alzheimer’s patient has high levels of PIB (red and yellow areas), \nwhereas the unaffected person’s brain has lower levels (blue).Image 1:\nCaption:The image shows two MRI images of the brain, one in blue and the other in yellow. The blue image shows the brain in detail, while the yellow image shows it in color. Both images are labeled with theit in color. Both images are labeled with the text \"Healthy person\" and \"Alzheimer's patient\".The image also includes a text that reads \"Figure 2.4B PET images of brains of a healthy person (left) and a person with Alzheimer's disease (right), Red and yellow colors indicate high levels of PIBRed and yellow colors indicate high levels of PIB bound to beta-amyloid plaques.\" This suggests that the image is related to the study of the human brain and the study's findings.Dangers of radioactive isotopes\nAlthough radioactive isotopes have many beneficial uses, uncontrolled exposure to high levels of radiation can be lethal. Theparticles and energy thrown off by radioactive atoms can damage molecules, especially DNA.Chernobyl nuclear disaster: Explosion of a nuclear reactor in Chernobyl, Ukraine, in 1986 released large amounts of radioactiveisotopes into the environment, which drifted over large areas of Russia, Belarus, and Europe. A few dozen people died fromacute radiation poisoning within a few weeks of the accident, and more than 100,000 people were evacuated from the area.Increased rates of thyroid cancer in children exposed to the radiation have been reported. 2011 post-tsunami Fukushima nuclear \ndisaster in Japan is another similar incident.disaster in Japan is another similar incident.\nNatural sources of radiation can also pose a threat. Radon, a radioactive gas, is \nthe second-leading cause of lung cancer in the United States.Radon can contaminate buildings in regions where underlying rocks naturally \ncontain uranium, a radioactive element. Homeowners can buy a radon detectoror hire a company to test their home to ensure that radon levels are safe. If \nlevels are found to be unsafe, technology exists to remove radon from home.Marie Curie (twice awarded Nobel Prize) died of aplastic pernicious anemia \n(body stops producing blood cells) due to her years of exposure to radiation \nduring her work.Image 1:\nCaption:The image is a diagram of the Uranium-238 Decay Chain, which is a periodic table that shows the atomic number of the elements in the periodic table. The diagram is divided into two columns, with theThe diagram is divided into two columns, with the left column representing the element name and the right column representing half-life units.The first column is labeled \"Only main decays are shown Gamma emitters are not indicated\" and shows the number of gamma emitters in each element. The second column is titled \"Element Names\" and listssecond column is titled \"Element Names\" and lists the elements that are represented in the diagram. The elements are arranged in a grid-like pattern, with each element represented by a square box.At the top of the diagram, there is a title that reads \"The Uranium -238 Decay Chain\" and below it, there are numbers that indicate the atomic decay chain. The atomic number is 82, 83, 84, 85, 86,chain. The atomic number is 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, and 92.Chemical bonds in biomolecules\nDistribution of electrons determine an atom’s chemical properties.\nChemical bond: When two atoms with incomplete outer shell electronsinteract, they loose, accept or share electrons to form a chemical bond.\n•\nIonic bond: Complete electron transfer between two atoms. Eg: NaCl\n••\nCovalent bond: Sharing of electrons between two atoms. Eg: H2, O2, CH4, \nH2O\n•\nHydrogen bonds: weak bonds between polar molecules; important in the \nchemistry of lifechemistry of life\nChemical reactions make and break chemical bonds\n2H2 + O2 2H2O \nAbove reaction is an explosive reaction (fortunately does not occur in cells)Chemical reactions essential for life on earth:\nPhotosynthesis: 6CO2 + 6H2O \nℎ𝑣C6H12O6 + 6O2\nAerobic respiration: C6H12O6 + 6O2 6CO2 + 6H2O + ATPImage 1:\nCaption:The image is a diagram that explains the overall equation for photosynthesis in plants. It shows the oxidation-reduction reaction in which H2O donates electrons (as hydrogen) for the reduction of CO2electrons (as hydrogen) for the reduction of CO2 to carbonhydrate (CH2O).The diagram is a flowchart that shows the process of photosynthesis, which is a process where photosynthesis takes place in the atmosphere of a plant. The diagram shows a sun with a carbon dioxideThe diagram shows a sun with a carbon dioxide molecule in the center, representing the energy of the sun. The molecule is represented by a yellow circle with a red line connecting it to the carbonwith a red line connecting it to the carbon dioxide molecules.At the bottom of the diagram, there is a text that explains that photosynthesis is the ultimate source of all biological en-ergy, photosynthetic organisms use the energy to sunlight to manufactureuse the energy to sunlight to manufacture glucose and other organic products, which heterotrophic cells use as energy and carbon sources.How did life originate on earth?\nTheory of spontaneous generation: If food was left to rot, it would generate living organisms like microbes,maggots (larvae of flies) and subsequently higher organisms like rats. But this was disproved by Louis Pasteur \nsaying ‘life only comes from life’.saying ‘life only comes from life’.\nDarwin’s theory of evolution suggested that simpler life forms gave rise to complex organisms.\nOparin-Haldane theory of abiogenesis (1929):Oparin-Haldane theory of abiogenesis (1929):\nOrganic molecules needed for life could be formed from abiotic \nmaterials in the presence of energy (lightning, heat and UV-light)under reducing conditions of early earth’s atmosphere (low levels of \nO2; major composition of CH4, NH3 and H2).\nOparin’s primordial soup: abiogenic factors above in water gave riseto biomolecules (amino acids, lipids, nucleic acids, carbohydrates) \nneeded to form cells, the primary constituents of life.\nAll above theories were only speculations without proof.Image 1:\nCaption: \nThe image is a diagram of a tree with different types of plants and animals. The tree is divided into different sections, each representing a different type of plant.The top section of the tree has a list of different types, including birds, reptiles, mammals, amphibians, land plants, and seaweed. There are also illustrations of animals such as a bird, a zebra, aof animals such as a bird, a zebra, a ladybug, and a turtle. The animals are arranged in a circular pattern around the tree, with the birds at the top and the reptiles at the bottom.On the left side of the image, there are illustrations of reptiles, reptiles and amphibians. On the right side, there is an illustration of a seaweed plant, which is a type of seaweed that is nativeplant, which is a type of seaweed that is native to the ocean floor. The seaweed is a reddish-brown color and has a long, thin trunk that extends from the top to the bottom of the trunk. There isfrom the top to the bottom of the trunk. There is also an illustration that shows a snail, a crab, a jellyfish, a starfish, and an octopus. The image is labeled with the names of the plants and theiris labeled with the names of the plants and their habitats.Miller – Urey experiment (1953)\nMiller simulated the origins of life on earth by using –\n•\nWater representing the oceans\n•\nReducing earth’s atmosphere consisting majorly of gases likeCH4, NH3 , H2 and H2O (water vapor)\n•\nEnergy source – sunlight, geothermal heat, electric spark for \nlightning/ thunder\nAfter a week, the water turned brown and containedcoacervates of organic biomolecules such as \n•\nAmino acids – building blocks for proteins\n•\nLipids – vesicles that can trap proteins and minerals in them \nto form cells\n•to form cells\n•\nSugars – building blocks for polysaccharides, glycoproteins, \netc.\nPrebiotic chemistry – amino acids, lipids and sugar are found onother planets, asteroids in our solar system and even beyond \nwithin our galaxy.\nIn one set of experiments, Miller included H2S from atmosphericvolcanoes, which accounted for formation of sulfur containing \namino acids and subsequent disulfide links in proteins.Image 1:\nCaption:The image is a diagram of an electrical spark (Lightning) system. It consists of a white sphere with a red arrow pointing towards it, representing the electrical spark. The sphere is connected to aelectrical spark. The sphere is connected to a blue line that runs from the top left corner of the image to the bottom right corner.There are two pipes connected to the sphere, one on the left side and the other on the right side. The pipes are connected to each other by a series of wires. The wires are labeled with the lettersof wires. The wires are labeled with the letters \"H2O, CH4, NH3, H2O\" and \"Lightning\".In the center of the diagram, there is a red circle with a blue wave-like pattern, representing a light source. On the bottom left corner, there are two purple pipes, one of which is connected with apurple pipes, one of which is connected with a purple pipe. The red circle represents the light source, while the blue circle represents a blue sphere.The diagram appears to be a representation of the process of electrical spark lightning, which is a type of lightning that is used to generate electricity from a source of light.Miller – Urey experiment proved the Oparin-Haldane chemical origin of life\nStep1\n• Free atoms of C, H, O, N\n• Abiotic synthesis of monomers from simple inorganic gases like H2, CH4, NH3 and H2O.• Monomers included amino acids, purines, pyrimidines, monosaccharides, fatty acids and glycerol\nStep2\n• Monomers combine to form polymers\n• Proteins, nucleic acids, lipids and polysaccharides\nStep3Step3\n• Formation of coacervates\n• Biomolecules surrounded by a lipid membrane\nStep4\n• Origin of nucleic acid and protein controlled metabolism inside a lipid membrane\n• Protocells or protobiontsWhich came first – DNA or Protein?\nFigure 1-36, A Possible RNA scenario world\nFirst cell probably used inorganic fuels produced by chemosynthesis for ATP \ngeneration: FeS + H2S ---FeS2 + H2generation: FeS + H2S ---FeS2 + H2\nFigure 1-37, Landmarks in the evolution of life on earth\nLehninger, Principles \nof Biochemistry, 6th Ed\nDevelopment of life stages on earth:\n•Development of life stages on earth:\n•\nChemical evolution: geological molecules \nto complex organic polymers\n•\nAcquisition of replicative ability: \ntransition from lifeless chemicallytransition from lifeless chemically \ninteracting entities to self replicating \nsystems\n•\nBiological evolution: eukaryotes, \nmulticellularity, plants and animals, \nmetabolic evolutionImage 1:\nCaption: \nThe image is a flowchart that explains the process of DNA replication. It is divided into six sections, each representing a different step in the process.The first section is titled \"Prebiotic formation of simple compounds, including nucleotides, from components of Earth's primitive atmosphere or gases in undersea volcanic vents\". The second sectionin undersea volcanic vents\". The second section is labeled \"Production of short RNA molecules with random sequences\". The third section is \"Selective replication of self-duplicating catalytic RNAreplication of self-duplicating catalytic RNA segments\". The fourth section is labelled \"Synthesis of specific peptides, catalyzed by RNA\". The fifth section is described as \"Increasing role offifth section is described as \"Increasing role of peptides in RNA replication, coevolution of RNA and protein\". The sixth section is called \"Primitive translation system develops, with RNA genome andtranslation system develops, with RNA genome and RNA-protein catalysts\".At the bottom of the flowchart, there is a label that reads \"Genomic RNA begins to be copied into DNA\". This indicates that the process is related to DNA genome translation and protein catalysts.Image 2:\nCaption:The image is a table that shows the different types of eukaryotes and their properties. The table is divided into two columns, with the left column representing the number of different species ofrepresenting the number of different species of organisms and the right column representing their characteristics.The first column is labeled \"Diversification of multicellular eukariesotes (plants, fungi, animals)\". The second column is titled \"Appearance of red and green algae\" and lists the characteristics ofand green algae\" and lists the characteristics of the species. The characteristics include:1,500, 2,000, and 3,000. The appearance of protists, the first euaryotes, and the development of O2-rich atmosphere\n\nThere are also three types of photosynthetic O2 producing cyanobacteriaAt the bottom of the table, there is a section titled \"Formation of oceans and continents\" which provides further information about the formation of Earth.Figure 1-38, Evolution of eukaryotes through endosymbiosis\nLehninger, Principles of Biochemistry, 6th Ed\nEvolution of eukaryotes through endosymbiosis\n•Evolution of eukaryotes through endosymbiosis\n•\nFew aerobic bacteria evolved into mitochondria of modern eukaryotes\n••\nPhotosynthetic cyanobacteria became plastids, chloroplasts of green algae, the ancestors of modern plant cells.Image 1:\nCaption: \nThe image is a diagram that shows the different types of chromosomes and their functions. It is divided into four sections, each representing a different type of chromosome.The first section on the left shows an anaerobic metabolism, which is a type of bacteria that is responsible for the formation of nuclei and nuclei. The nuclei are responsible for storing nuclei inThe nuclei are responsible for storing nuclei in the nucleus of the cell. The nucleus is then released into the cell, which then releases nuclei into the mitochondria. The mitochondria is theninto the mitochondria. The mitochondria is then converted into a nuclei, which helps to store and transport the nuclei to the nucleus. The image also shows the mitochondrion, which acts as a catalystshows the mitochondrion, which acts as a catalyst for the cell's growth and development.In the center of the image, there are four chromosomes, each with a different color - blue, green, and white. These chromosomes are arranged in a circular pattern, representing the different stagespattern, representing the different stages of the cycle. The blue chromosomes represent the nucleis, while the green chromosomes represent photosynthetic chromosomes. The white chromosomes representchromosomes. The white chromosomes represent chloroplast, which are the cells that make up the cell membrane. The diagram also includes text that explains the process of photosynthesis and howexplains the process of photosynthesis and how photosynthesis can be used to convert photosynthesis into photosynthesis.\n","output_type":"stream"}]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]}]}